---
output:
  md_document:
    variant: markdown_github
---

```{r, echo = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  fig.path = "README-"
)
```

# metacell

The MetaCell package facilitate analysis of single cell RNA-seq UMI matrices through generation of cell-to-cell similarity graphs and computation of partitions of the graph into small (~20-200 typically) homogeneous groups of cells. The derived metacells are then used for deriving different representations of the data, and form the basis for analysis of cell types, subtypes, transcriptional gradients, cell-cycle variation, gene modules and their regulatory models and more. More details on the usage of the MetaCell pipeline is available in the package vignettes, and in papers using it.

References:

Method: Baran et al. 2018 (bioarxiv link comming up)

Examples of applications:

Giladi et al., Nat Cell Biol 2018 (Mouse hematopoiesis scRNA-seq)
Sebe-pedros et al. NEE 2018 (whole organisms scRNA-seq)
Sebe-pedros et al., Cell 2018 (whole organism scRNA-seq)

## Installation

```{r, eval=FALSE}
install.packages('metacell', repos=c(getOption('repos'), 'https://tanaylab.bitbucket.io/repo'))
```


