<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Supervised filtering of feature genes: guided tutorial on 8K PBMCs • metacell</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><!-- docsearch --><script src="../docsearch.js"></script><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.css" integrity="sha256-QOSRU/ra9ActyXkIBbiIB144aDBdtvXBcNc3OTNuX/Q=" crossorigin="anonymous">
<link href="../docsearch.css" rel="stylesheet">
<script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/jquery.mark.min.js" integrity="sha256-4HLtjeVgH0eIB3aZ9mLYF6E8oU5chNdjU6p6rrXpl9U=" crossorigin="anonymous"></script><meta property="og:title" content="Supervised filtering of feature genes: guided tutorial on 8K PBMCs">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">metacell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.41</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/basic_pbmc8k.html">Usage</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/a-basic_pbmc8k.html">Running metacell analysis: guided tutorial on 8K PBMCs</a>
    </li>
    <li>
      <a href="../articles/b-geneset_by_anchor_pbmc8k.html">Supervised filtering of feature genes: guided tutorial on 8K PBMCs</a>
    </li>
    <li>
      <a href="../articles/c-metacell_annotation.html">Annotation of metacells</a>
    </li>
    <li>
      <a href="../articles/d-amphimedon.html">Analyzing whole-organism scRNA-seq data with metacell</a>
    </li>
    <li>
      <a href="../articles/e-data_convert.html">Interoperability with other scRNA formats</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a></a>
</li>
      </ul>
<form class="navbar-form navbar-right" role="search">
        <div class="form-group">
          <input type="search" class="form-control" name="search-input" id="search-input" placeholder="Search..." aria-label="Search for..." autocomplete="off">
</div>
      </form>
      
    </div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Supervised filtering of feature genes: guided tutorial on 8K PBMCs</h1>
                        <h4 class="author">Yaniv Lubling</h4>
            
            <h4 class="date">2018-10-10</h4>
      
      
      <div class="hidden name"><code>b-geneset_by_anchor_pbmc8k.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>This note walks through the common scenario of disabling specific gene modules from affecting the partitioning of cells into metacells. In many applications we would want to mask technical effects of stress (e.g. heat shock) from affecting the partitioning, or to mask cell cycle related genes in order to get metacells that represent cell types/states and not stages at the cell cycle.</p>
<p>We’ll start with finding genes that are correlated to a few cell cycle and stress halmark genes, then generate gene correlation heatmaps that will help to further filter the genes, and lastly show how to create metacells that are not affected by these genes.</p>
<p>We use here the small 8K PBMC dataset that is used in the <strong>Running metacell analysis: guided tutorial on 8K PBMCs</strong> vignette, please make sure to run that vignette before executing this one, since we’ll use several objects produced there.</p>
</div>
<div id="setting-up" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up" class="anchor"></a>Setting up</h2>
<p>We start by loading the library and initializating the database and the figures directories:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(<span class="st">"metacell"</span>)

<span class="kw"><a href="../reference/scdb_init.html">scdb_init</a></span>(<span class="st">"testdb/"</span>, <span class="dt">force_reinit=</span>T)
<span class="co">#&gt; initializing scdb to testdb/</span>

figs_dir =<span class="st"> "figs"</span>
<span class="kw"><a href="../reference/scfigs_init.html">scfigs_init</a></span>(figs_dir)</code></pre></div>
</div>
<div id="creating-the-gene-correlation-matrix" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-the-gene-correlation-matrix" class="anchor"></a>Creating the gene correlation matrix</h2>
<p>We’re going to generate a gene-gene correlation matrix using the umi matrix generated in the <strong>Running metacell analysis: guided tutorial on 8K PBMCs</strong> vignette. The genes will then be filtered by thresholding their correlation with the user supplied genes (termed anchor genes). In this vignette we’ll build a set of cell cycle and stress related genes. So let’s define these gene anchors, and some additional variables:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes_anchors =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="st">'MKI67'</span>, <span class="st">'PCNA'</span>, <span class="st">'TOP2A'</span>, <span class="st">'TXN'</span>, <span class="st">'HSP90AB1'</span>, <span class="st">'FOS'</span>)
tab_fn =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(figs_dir, <span class="st">"lateral_gmods.txt"</span>, <span class="dt">sep=</span><span class="st">"/"</span>)
gset_nm =<span class="st"> "lateral"</span></code></pre></div>
<p><em>tab_fn</em> is the output file name of the correlation matrix, and <em>gset_nm</em> is the id of the gene set that will be generated.</p>
<p>Now we’ll call the function that computes the gene-gene correlation matrix:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mcell_mat_rpt_cor_anchors.html">mcell_mat_rpt_cor_anchors</a></span>(<span class="dt">mat_id=</span><span class="st">"test"</span>, <span class="dt">gene_anchors =</span> genes_anchors, <span class="dt">cor_thresh =</span> <span class="fl">0.1</span>, <span class="dt">gene_anti =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(), <span class="dt">tab_fn =</span> tab_fn, <span class="dt">sz_cor_thresh =</span> <span class="fl">0.2</span>)</code></pre></div>
<p>The function computes the correlations of all genes in the umi matrix with the gene anchors after downsampling the umi counts. It then filters genes that their correlation to the anchor genes is above <em>cor_thresh</em>. Genes can also be filtered based on their correlation to a set of negative control genes, that can be specified in the <em>gene_anti</em> parameter. In this case, genes that their top correlated gene is a negative control and not one of the anchor genes are discarded. If the <em>sz_cor_thresh</em> is not <em>NA</em>, genes with correlation to the cells total umi counts above <em>sz_cor_thresh</em> are also selected. In our case, we expect some cell cycle genes to be correlated with cell size, so we specify a positive threshold.</p>
<p>Let’s read the filtered gene correlation matrix and examine it:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gcor_mat =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(tab_fn, <span class="dt">header=</span>T)
<span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/head">head</a></span>(gcor_mat)
<span class="co">#&gt;                  sz_cor        MKI67          PCNA        TOP2A        TXN</span>
<span class="co">#&gt; ABCB9       0.002267035  0.113515805  0.0077224535  0.060206640 0.01030069</span>
<span class="co">#&gt; ABHD5      -0.004551788 -0.006990550 -0.0052722943 -0.007749114 0.04990038</span>
<span class="co">#&gt; AC098614.2  0.007561982  0.072206551  0.0077146734  0.132693409 0.01243514</span>
<span class="co">#&gt; ACSL1      -0.015851397 -0.005267215 -0.0008923261 -0.005838775 0.03315487</span>
<span class="co">#&gt; ACTB        0.253784528  0.108718458  0.0454909890  0.101454370 0.15426727</span>
<span class="co">#&gt; ACTG1       0.221359303  0.110300254  0.0220282723  0.108523623 0.04266898</span>
<span class="co">#&gt;                HSP90AB1           FOS       max neg_max</span>
<span class="co">#&gt; ABCB9       0.006374946 -0.0080735106 0.1135158       0</span>
<span class="co">#&gt; ABHD5      -0.016661139  0.1031994092 0.1031994       0</span>
<span class="co">#&gt; AC098614.2 -0.002731917 -0.0004316086 0.1326934       0</span>
<span class="co">#&gt; ACSL1      -0.049008246  0.1054241285 0.1054241       0</span>
<span class="co">#&gt; ACTB       -0.068296466  0.1408468643 0.1542673       0</span>
<span class="co">#&gt; ACTG1       0.030165625 -0.0957211702 0.1103003       0</span>
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/dim">dim</a></span>(gcor_mat))
<span class="co">#&gt; [1] 582   9</span></code></pre></div>
<p>The row names are genes, and the columns are the correlation to the total umi counts (<em>sz_cor</em>), the correlations to the anchor genes and their maximal one (<em>max</em>). Since we didn’t specify negative anchor genes in this example the maximal correlation to the negative anchors is zero (<em>max_neg</em>).</p>
<p>Now we’ll add a gene set to the database, comprised of the genes in the filtered matrix:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">foc_genes =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/apply">apply</a></span>(gcor_mat[, genes_anchors], <span class="dv">1</span>, which.max)

gset =<span class="st"> </span><span class="kw"><a href="../reference/gset_new_gset.html">gset_new_gset</a></span>(<span class="dt">sets =</span> foc_genes, <span class="dt">desc =</span> <span class="st">"Cell cycle and stress correlated genes"</span>)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(gset)
<span class="co">#&gt; An object of class "tgGeneSets"</span>
<span class="co">#&gt; Slot "description":</span>
<span class="co">#&gt; [1] "Cell cycle and stress correlated genes"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "set_names":</span>
<span class="co">#&gt; [1] 1 6 3 4 2 5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "gene_set":</span>
<span class="co">#&gt;         ABCB9         ABHD5    AC098614.2         ACSL1          ACTB </span>
<span class="co">#&gt;             1             6             3             6             4 </span>
<span class="co">#&gt;         ACTG1         ACTR2        ADGRF3        AGTRAP         AHNAK </span>
<span class="co">#&gt;             1             6             3             6             6 </span>
<span class="co">#&gt;           AHR          AIF1       ALDH1A1         ALDH2       ALDH3B1 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;         ALDOA       ALOX5AP       ANKRD28         ANPEP         ANXA1 </span>
<span class="co">#&gt;             6             4             6             6             6 </span>
<span class="co">#&gt;         ANXA2         ANXA5          AOAH         AP1S2         AP2S1 </span>
<span class="co">#&gt;             4             6             6             6             6 </span>
<span class="co">#&gt;         APLP2           APP     ARHGAP11A       ARHGDIB      ARHGEF40 </span>
<span class="co">#&gt;             6             4             1             1             6 </span>
<span class="co">#&gt;         ARL5B        ARRDC4         ASAH1         ASF1B         ASGR1 </span>
<span class="co">#&gt;             6             6             6             1             6 </span>
<span class="co">#&gt;          ASIP          ATF3          ATG3         ATOX1    ATP2B1-AS1 </span>
<span class="co">#&gt;             4             6             6             6             6 </span>
<span class="co">#&gt;         ATP5E        ATP5G2         ATP5I         ATP5L       ATP6V0B </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;      ATP6V0E1       ATP6V1A      ATP6V1B2       ATP6V1F         AURKA </span>
<span class="co">#&gt;             6             6             6             6             3 </span>
<span class="co">#&gt;         BACH1        BCL2A1         BIRC5       BLOC1S1         BLVRB </span>
<span class="co">#&gt;             6             6             1             6             6 </span>
<span class="co">#&gt;        BNIP3L          BRI3          BST1      C10orf11      C12orf75 </span>
<span class="co">#&gt;             6             6             6             6             4 </span>
<span class="co">#&gt;       C14orf2      C19orf38      C1orf162       C4orf48         C5AR1 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;       C9orf72         CALM2          CAPG        CAPZA2        CARD16 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;         CARS2         CASP1         CASZ1      CATSPER1        CCDC50 </span>
<span class="co">#&gt;             6             6             3             6             4 </span>
<span class="co">#&gt;       CCDC88A          CCL3        CCL3L3         CCNA2         CCNB1 </span>
<span class="co">#&gt;             6             6             6             3             1 </span>
<span class="co">#&gt;         CCNL1          CCNY          CCR9          CD14         CD163 </span>
<span class="co">#&gt;             6             6             3             6             6 </span>
<span class="co">#&gt;          CD1D        CD300E       CD300LF         CD302          CD33 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;          CD36           CD4          CD44          CD63          CD83 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;          CD86          CD93           CDA         CDCA5         CDCA8 </span>
<span class="co">#&gt;             6             6             6             1             1 </span>
<span class="co">#&gt;        CDKN2C         CDKN3          CDT1         CEBPB         CEBPD </span>
<span class="co">#&gt;             3             1             1             6             6 </span>
<span class="co">#&gt;         CECR1         CENPF         CENPK           CFD           CFP </span>
<span class="co">#&gt;             6             1             3             6             6 </span>
<span class="co">#&gt;         CHEK1        CHMP1B          CHP1          CIB2         CKAP2 </span>
<span class="co">#&gt;             3             6             6             4             3 </span>
<span class="co">#&gt;         CKAP4         CKS1B          CKS2       CLEC10A       CLEC12A </span>
<span class="co">#&gt;             6             3             1             4             6 </span>
<span class="co">#&gt;        CLEC4A        CLEC4C        CLEC4D        CLEC4E        CLEC7A </span>
<span class="co">#&gt;             6             4             6             6             6 </span>
<span class="co">#&gt;         CLIC2         CMTM6         CNPY3         COTL1         COX5B </span>
<span class="co">#&gt;             4             6             4             6             6 </span>
<span class="co">#&gt;        COX6B1         COX7B        CPPED1          CPVL         CREB5 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;         CREG1         CRTAP        CSF2RB         CSF3R          CST3 </span>
<span class="co">#&gt;             6             6             4             6             6 </span>
<span class="co">#&gt;          CSTA    CTB-61M7.2  CTC-510F12.4  CTD-3252C9.4          CTSB </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;          CTSC          CTSD          CTSH          CTSS          CTSZ </span>
<span class="co">#&gt;             4             6             6             6             6 </span>
<span class="co">#&gt;         CXCL8          CYBA          CYBB        CYP1B1        DAZAP2 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;         DDIAS         DERL3         DMXL2      DNASE1L3          DPYD </span>
<span class="co">#&gt;             1             4             6             4             6 </span>
<span class="co">#&gt;        DPYSL2         DSCC1         DUSP1         DUSP6           DUT </span>
<span class="co">#&gt;             6             1             6             6             1 </span>
<span class="co">#&gt;        DYNLT1          E2F1          ECT2     EEF1A1P42       EHBP1L1 </span>
<span class="co">#&gt;             6             2             1             3             6 </span>
<span class="co">#&gt;          EIF1      EIF4EBP1          ELOB          ENHO        ENTPD1 </span>
<span class="co">#&gt;             6             6             6             4             6 </span>
<span class="co">#&gt;          EREG         EVI2A         EVI2B          EZH2         F13A1 </span>
<span class="co">#&gt;             6             6             6             1             6 </span>
<span class="co">#&gt;       FAM120A       FAM198B        FAM45A         FANCI         FBXL5 </span>
<span class="co">#&gt;             6             6             6             1             6 </span>
<span class="co">#&gt;         FBXO5          FCAR        FCER1A        FCER1G        FCGR1A </span>
<span class="co">#&gt;             1             6             4             6             6 </span>
<span class="co">#&gt;        FCGR2A         FCGRT          FCN1          FGD4          FGL2 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;           FGR        FKBP1A          FLT3         FOLR3           FOS </span>
<span class="co">#&gt;             6             6             4             6             6 </span>
<span class="co">#&gt;          FOSB         FOSL2          FPR1          FTH1           FTL </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;         FTLP3          G0S2     GABARAPL1         GAPDH          GAS6 </span>
<span class="co">#&gt;             6             6             6             6             4 </span>
<span class="co">#&gt;          GAS7           GCA         GINS2        GLIPR1          GLMP </span>
<span class="co">#&gt;             6             6             2             6             3 </span>
<span class="co">#&gt;          GLRX        GLT1D1          GLUL          GMFG         GNAI2 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;          GNAQ           GNS        GPBAR1        GPCPD1          GPX1 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;          GPX4           GRN           GSN         GSTO1         GSTP1 </span>
<span class="co">#&gt;             6             6             4             6             6 </span>
<span class="co">#&gt;          GZMB         H2AFY         H2AFZ         H3F3B         HACD4 </span>
<span class="co">#&gt;             4             6             1             6             6 </span>
<span class="co">#&gt;           HCK         HIPK3      HIST1H4C           HK2      HLA-DPA1 </span>
<span class="co">#&gt;             6             6             3             6             4 </span>
<span class="co">#&gt;      HLA-DPB1      HLA-DQA1       HLA-DRA      HLA-DRB1      HLA-DRB6 </span>
<span class="co">#&gt;             4             4             4             4             4 </span>
<span class="co">#&gt;         HMGB2          HNMT         HSBP1      HSP90AB1         HSPA8 </span>
<span class="co">#&gt;             1             6             6             5             5 </span>
<span class="co">#&gt;           ID1          IER2          IER3          IER5         IFI30 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;        IFITM3        IFNGR1        IFNGR2         IGSF6       IL13RA1 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;          IL1B         IL1RN        IQGAP1         IRAK3       IRF2BP2 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;          IRF7          IRF8          IRS2         ITGAM         ITM2B </span>
<span class="co">#&gt;             4             4             6             6             6 </span>
<span class="co">#&gt;         ITM2C          JAML           JUN          JUNB          JUND </span>
<span class="co">#&gt;             4             6             6             6             6 </span>
<span class="co">#&gt;         KCNE3        KCTD12      KIAA0930        KIF18A         KIF23 </span>
<span class="co">#&gt;             6             6             6             1             3 </span>
<span class="co">#&gt;         KIFC1         KLF10          KLF4          KLF6         KPNA2 </span>
<span class="co">#&gt;             1             6             6             6             3 </span>
<span class="co">#&gt;         LACTB         LAMP2         LAMP5       LAMTOR2       LAMTOR4 </span>
<span class="co">#&gt;             6             6             4             6             6 </span>
<span class="co">#&gt;          LCP1        LGALS1        LGALS2        LGALS3        LILRA2 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;        LILRA4        LILRA5        LILRB2        LILRB3        LILRB4 </span>
<span class="co">#&gt;             4             6             6             6             4 </span>
<span class="co">#&gt;     LINC00877     LINC00996          LMO2        LPCAT2          LRP1 </span>
<span class="co">#&gt;             6             4             6             6             6 </span>
<span class="co">#&gt;        LRRC25        LRRC26         LRRK2          LSM6          LST1 </span>
<span class="co">#&gt;             6             4             6             6             6 </span>
<span class="co">#&gt;         LTA4H          LY96          LYST           LYZ        MAD2L1 </span>
<span class="co">#&gt;             6             6             6             6             1 </span>
<span class="co">#&gt;          MAFB         MAP1A        MARCKS        MCEMP1          MCL1 </span>
<span class="co">#&gt;             6             4             6             6             6 </span>
<span class="co">#&gt;          MCM7         MEGF9        METTL9         MFSD1         MGST1 </span>
<span class="co">#&gt;             2             6             6             6             6 </span>
<span class="co">#&gt;         MGST2          MIDN         MKI67          MNDA         MPEG1 </span>
<span class="co">#&gt;             6             6             1             6             6 </span>
<span class="co">#&gt;        MS4A6A         MS4A7         MSRB1         MYADM         MYBL2 </span>
<span class="co">#&gt;             6             6             6             6             4 </span>
<span class="co">#&gt;          MYL6          MZB1         NAMPT         NAPRT         NAPSB </span>
<span class="co">#&gt;             4             4             6             6             4 </span>
<span class="co">#&gt;          NCF2       NDUFA11        NDUFB1        NDUFS6          NEFH </span>
<span class="co">#&gt;             6             6             6             6             3 </span>
<span class="co">#&gt;         NFAM1          NFE2        NFKBIA        NFKBIZ         NINJ1 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;         NLRP3          NPC2          NPM1         NR4A2          NRG1 </span>
<span class="co">#&gt;             6             4             5             6             6 </span>
<span class="co">#&gt;        NUDT16          NUF2        NUP214        NUSAP1          OAZ1 </span>
<span class="co">#&gt;             6             1             6             1             6 </span>
<span class="co">#&gt;         ODF3B        OGFRL1         OSCAR         OTUD1        P2RY13 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;       PACSIN1          PAK1         PCBP1         PCLAF          PCNA </span>
<span class="co">#&gt;             4             6             6             1             2 </span>
<span class="co">#&gt;         PFDN5          PFN1           PGD          PGLS          PHEX </span>
<span class="co">#&gt;             6             4             6             6             4 </span>
<span class="co">#&gt;        PICALM          PID1         PILRA        PLA2G7         PLAUR </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;         PLBD1          PLD4          PLEK          PLP2        PLSCR1 </span>
<span class="co">#&gt;             6             4             4             4             6 </span>
<span class="co">#&gt;        PLXDC2        PLXNB2         PNRC1         POLE4        POLR2L </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;          POMP          PPIB      PPP1R14B      PPP1R15A         PRAM1 </span>
<span class="co">#&gt;             6             4             4             6             6 </span>
<span class="co">#&gt;         PRDX3       PRELID1          PROC          PSAP         PSMB3 </span>
<span class="co">#&gt;             6             6             4             6             6 </span>
<span class="co">#&gt;         PTAFR         PTGDS         PTGS2          PTMS         PTPRE </span>
<span class="co">#&gt;             6             4             6             4             6 </span>
<span class="co">#&gt;         PTTG1        PYCARD          PYGL           QKI          QPCT </span>
<span class="co">#&gt;             1             6             6             6             6 </span>
<span class="co">#&gt;     RAB11FIP1         RAB31         RAB32         RAB3D          RAC1 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;        RASSF2        RASSF4         RBM47          RBP7        RCBTB2 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;          RETN          RGS2          RHOA          RHOB          RHOG </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;         RIPK2     RN7SL368P     RNA5SP151        RNASE2        RNASE6 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;        RNF130        RNF181        RNF19B      RNU4ATAC      RNVU1-19 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt; RP11-1143G9.4  RP11-160E2.6 RP11-363K21.1   RP11-40C6.2 RP11-522I20.3 </span>
<span class="co">#&gt;             6             6             6             5             3 </span>
<span class="co">#&gt;  RP11-700H6.1  RP11-73G16.2        RPL10A        RPL23A        RPL27A </span>
<span class="co">#&gt;             1             4             5             5             5 </span>
<span class="co">#&gt;          RPL3         RPL30         RPL31          RPL5         RPLP0 </span>
<span class="co">#&gt;             5             5             5             5             5 </span>
<span class="co">#&gt;         RPLP2         RPS12         RPS18          RPS2         RPS24 </span>
<span class="co">#&gt;             5             5             5             5             6 </span>
<span class="co">#&gt;         RPS27        RPS27A         RPS29          RPS3         RPS4X </span>
<span class="co">#&gt;             5             5             5             5             5 </span>
<span class="co">#&gt;          RPS5          RPS6          RPSA          RRM2          RTN3 </span>
<span class="co">#&gt;             5             5             5             3             6 </span>
<span class="co">#&gt;          RTN4          RXRA       S100A10       S100A11       S100A12 </span>
<span class="co">#&gt;             6             6             4             6             6 </span>
<span class="co">#&gt;        S100A4        S100A6        S100A8        S100A9        SAMHD1 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;         SAP30          SAT1        SCAMP5        SCPEP1           SCT </span>
<span class="co">#&gt;             3             6             4             6             4 </span>
<span class="co">#&gt;         SDCBP        SEC11A        SEC61B        SEC61G        SEMA4A </span>
<span class="co">#&gt;             6             6             4             4             6 </span>
<span class="co">#&gt;         SERF2         SERP1      SERPINA1      SERPINF1          SGK1 </span>
<span class="co">#&gt;             6             6             6             4             6 </span>
<span class="co">#&gt;          SGO1          SGO2      SH3BGRL3         SIRPA        SIRPB1 </span>
<span class="co">#&gt;             1             1             4             6             6 </span>
<span class="co">#&gt;       SLC11A1       SLC15A3       SLC15A4       SLC16A3       SLC31A2 </span>
<span class="co">#&gt;             6             6             4             6             6 </span>
<span class="co">#&gt;        SLC7A7        SLC8A1         SMCO4        SMIM25         SMIM5 </span>
<span class="co">#&gt;             6             6             6             6             4 </span>
<span class="co">#&gt;         SMPD3          SNCA     SNORD3B-2         SNX10          SOD2 </span>
<span class="co">#&gt;             4             4             6             6             6 </span>
<span class="co">#&gt;         SPCS1         SPG21          SPI1          SRGN        SRPK2P </span>
<span class="co">#&gt;             4             6             6             6             1 </span>
<span class="co">#&gt;         STAB1         STMN1         STX11        STXBP2         SULF2 </span>
<span class="co">#&gt;             6             1             6             6             6 </span>
<span class="co">#&gt;       SULT1A1           SYK        TALDO1          TCF4          TET2 </span>
<span class="co">#&gt;             6             6             6             4             6 </span>
<span class="co">#&gt;          TFEC         TGFBI        TIMM8B         TIMP1         TIMP2 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;           TK1           TKT          TLR2          TLR8       TMEM107 </span>
<span class="co">#&gt;             3             6             6             6             6 </span>
<span class="co">#&gt;      TMEM167A      TMEM170B       TMEM205       TNFAIP2       TNFAIP3 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;      TNFRSF1B      TNFRSF21       TNFSF10      TNFSF13B         TONSL </span>
<span class="co">#&gt;             6             4             6             6             1 </span>
<span class="co">#&gt;         TOP2A          TPI1          TPM2          TPP1          TPX2 </span>
<span class="co">#&gt;             3             6             4             6             3 </span>
<span class="co">#&gt;         TREM1         TRIB1          TSPO        TUBA1B        TUBA1C </span>
<span class="co">#&gt;             6             6             6             3             1 </span>
<span class="co">#&gt;          TUBB        TUBB4B           TXN          TYMP          TYMS </span>
<span class="co">#&gt;             1             1             4             6             1 </span>
<span class="co">#&gt;        TYROBP          UGCG       UNC93B1        UQCR11          USP1 </span>
<span class="co">#&gt;             6             4             4             6             3 </span>
<span class="co">#&gt;         VAMP8          VCAN           VIM          VSIR         VSTM1 </span>
<span class="co">#&gt;             4             6             6             6             6 </span>
<span class="co">#&gt;          WARS          YBX3         YPEL5         YWHAE          ZEB2 </span>
<span class="co">#&gt;             6             6             6             6             6 </span>
<span class="co">#&gt;        ZFAND5         ZFP36       ZFP36L1       ZNF385A         ZNF81 </span>
<span class="co">#&gt;             6             6             6             6             1 </span>
<span class="co">#&gt;        ZSWIM6         ZWINT </span>
<span class="co">#&gt;             6             3</span>

<span class="kw"><a href="../reference/scdb_add_gset.html">scdb_add_gset</a></span>(gset_nm, gset)</code></pre></div>
<p>Note that we constructed a membership vector with genes as names and the number of the top correlated anchor gene as the value. This is optional, we could have created a gene set with all genes in the same set (e.g. 1). It is important to add the gene set to the database for the next steps. We use <em>gset_nm</em> as the gene set id.</p>
</div>
<div id="fine-clustering-and-filtering-of-the-gene-set" class="section level2">
<h2 class="hasAnchor">
<a href="#fine-clustering-and-filtering-of-the-gene-set" class="anchor"></a>Fine clustering and filtering of the gene set</h2>
<p>To further examine the gene set we got, we’ll cluster the genes. First, we’ll create a sub-matrix from the umi matrix that will contain only the genes in the gene set:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">sub_mat_id =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste</a></span>(<span class="st">"test"</span>, gset_nm, <span class="dt">sep=</span><span class="st">"_"</span>)
<span class="kw"><a href="../reference/mcell_mat_ignore_genes.html">mcell_mat_ignore_genes</a></span>(<span class="dt">new_mat_id =</span> sub_mat_id, <span class="dt">mat_id =</span> <span class="st">"test"</span>, <span class="dt">ig_genes =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/names">names</a></span>(foc_genes), <span class="dt">reverse =</span> T)</code></pre></div>
<p>Now we’ll cluster the genes on the downsampled sub-matrix, using hierarchical clustering and cutting the tree to 20 clusters:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mcell_gset_split_by_dsmat.html">mcell_gset_split_by_dsmat</a></span>(<span class="dt">gset_id =</span> gset_nm, <span class="dt">mat_id =</span> sub_mat_id, <span class="dt">K =</span> <span class="dv">20</span>)
<span class="co">#&gt; will downsample the matrix, N= 750</span>
<span class="co">#&gt; 100%</span></code></pre></div>
<p>The number of clusters should be proportional to the number of genes we cluster, as we want to get homogeneous clusters of coodinated genes.</p>
<p>The cluster membership information is updated in the gene set. Let’s re-load it and have a look:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">gset =<span class="st"> </span><span class="kw"><a href="../reference/scdb_gset.html">scdb_gset</a></span>(gset_nm)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(gset)
<span class="co">#&gt; An object of class "tgGeneSets"</span>
<span class="co">#&gt; Slot "description":</span>
<span class="co">#&gt; [1] "Cell cycle and stress correlated genes hc K=20"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "set_names":</span>
<span class="co">#&gt;  [1]  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19 20</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "gene_set":</span>
<span class="co">#&gt;         ABCB9         ABHD5    AC098614.2         ACSL1          ACTB </span>
<span class="co">#&gt;             1             2             3             2             4 </span>
<span class="co">#&gt;         ACTG1         ACTR2        ADGRF3        AGTRAP         AHNAK </span>
<span class="co">#&gt;             5             6             3             7             4 </span>
<span class="co">#&gt;           AHR          AIF1       ALDH1A1         ALDH2       ALDH3B1 </span>
<span class="co">#&gt;             2             8             7             9             2 </span>
<span class="co">#&gt;         ALDOA       ALOX5AP       ANKRD28         ANPEP         ANXA1 </span>
<span class="co">#&gt;             4            10             2            11             4 </span>
<span class="co">#&gt;         ANXA2         ANXA5          AOAH         AP1S2         AP2S1 </span>
<span class="co">#&gt;             4             4             2            12             9 </span>
<span class="co">#&gt;         APLP2           APP     ARHGAP11A       ARHGDIB      ARHGEF40 </span>
<span class="co">#&gt;             7            13            14             5             2 </span>
<span class="co">#&gt;         ARL5B        ARRDC4         ASAH1         ASF1B         ASGR1 </span>
<span class="co">#&gt;             2             7            11            14            15 </span>
<span class="co">#&gt;          ASIP          ATF3          ATG3         ATOX1    ATP2B1-AS1 </span>
<span class="co">#&gt;            13             2             4             6            15 </span>
<span class="co">#&gt;         ATP5E        ATP5G2         ATP5I         ATP5L       ATP6V0B </span>
<span class="co">#&gt;             6             3             6             3             9 </span>
<span class="co">#&gt;      ATP6V0E1       ATP6V1A      ATP6V1B2       ATP6V1F         AURKA </span>
<span class="co">#&gt;             1             2             2             6             3 </span>
<span class="co">#&gt;         BACH1        BCL2A1         BIRC5       BLOC1S1         BLVRB </span>
<span class="co">#&gt;             2             2            16             2            12 </span>
<span class="co">#&gt;        BNIP3L          BRI3          BST1      C10orf11      C12orf75 </span>
<span class="co">#&gt;             2            11             7             2            10 </span>
<span class="co">#&gt;       C14orf2      C19orf38      C1orf162       C4orf48         C5AR1 </span>
<span class="co">#&gt;             7            11             4             7            11 </span>
<span class="co">#&gt;       C9orf72         CALM2          CAPG        CAPZA2        CARD16 </span>
<span class="co">#&gt;             2             4             9             6            11 </span>
<span class="co">#&gt;         CARS2         CASP1         CASZ1      CATSPER1        CCDC50 </span>
<span class="co">#&gt;             2            11             3             7            13 </span>
<span class="co">#&gt;       CCDC88A          CCL3        CCL3L3         CCNA2         CCNB1 </span>
<span class="co">#&gt;             9            15            15            16            16 </span>
<span class="co">#&gt;         CCNL1          CCNY          CCR9          CD14         CD163 </span>
<span class="co">#&gt;            17             2             3            12             7 </span>
<span class="co">#&gt;          CD1D        CD300E       CD300LF         CD302          CD33 </span>
<span class="co">#&gt;             2            11            11             9             2 </span>
<span class="co">#&gt;          CD36           CD4          CD44          CD63          CD83 </span>
<span class="co">#&gt;            12             9             3             6            15 </span>
<span class="co">#&gt;          CD86          CD93           CDA         CDCA5         CDCA8 </span>
<span class="co">#&gt;             2             7            11             1            16 </span>
<span class="co">#&gt;        CDKN2C         CDKN3          CDT1         CEBPB         CEBPD </span>
<span class="co">#&gt;             3            16            14            11             9 </span>
<span class="co">#&gt;         CECR1         CENPF         CENPK           CFD           CFP </span>
<span class="co">#&gt;             7            16            14             8             8 </span>
<span class="co">#&gt;         CHEK1        CHMP1B          CHP1          CIB2         CKAP2 </span>
<span class="co">#&gt;            14             1             2            13             3 </span>
<span class="co">#&gt;         CKAP4         CKS1B          CKS2       CLEC10A       CLEC12A </span>
<span class="co">#&gt;             7             3             3            18            11 </span>
<span class="co">#&gt;        CLEC4A        CLEC4C        CLEC4D        CLEC4E        CLEC7A </span>
<span class="co">#&gt;             9            13             7             7             8 </span>
<span class="co">#&gt;         CLIC2         CMTM6         CNPY3         COTL1         COX5B </span>
<span class="co">#&gt;            18             2             2             4             1 </span>
<span class="co">#&gt;        COX6B1         COX7B        CPPED1          CPVL         CREB5 </span>
<span class="co">#&gt;             3             3             9             8             2 </span>
<span class="co">#&gt;         CREG1         CRTAP        CSF2RB         CSF3R          CST3 </span>
<span class="co">#&gt;             2             2             9            12             8 </span>
<span class="co">#&gt;          CSTA    CTB-61M7.2  CTC-510F12.4  CTD-3252C9.4          CTSB </span>
<span class="co">#&gt;             8             7             7            15             9 </span>
<span class="co">#&gt;          CTSC          CTSD          CTSH          CTSS          CTSZ </span>
<span class="co">#&gt;             6            12             9             8             9 </span>
<span class="co">#&gt;         CXCL8          CYBA          CYBB        CYP1B1        DAZAP2 </span>
<span class="co">#&gt;            15             9             8             7             1 </span>
<span class="co">#&gt;         DDIAS         DERL3         DMXL2      DNASE1L3          DPYD </span>
<span class="co">#&gt;             1            13             2            13             1 </span>
<span class="co">#&gt;        DPYSL2         DSCC1         DUSP1         DUSP6           DUT </span>
<span class="co">#&gt;             9             3            17            15             3 </span>
<span class="co">#&gt;        DYNLT1          E2F1          ECT2     EEF1A1P42       EHBP1L1 </span>
<span class="co">#&gt;             2            14             3             3             2 </span>
<span class="co">#&gt;          EIF1      EIF4EBP1          ELOB          ENHO        ENTPD1 </span>
<span class="co">#&gt;            17            15             7            18             2 </span>
<span class="co">#&gt;          EREG         EVI2A         EVI2B          EZH2         F13A1 </span>
<span class="co">#&gt;             2             1             2             3             1 </span>
<span class="co">#&gt;       FAM120A       FAM198B        FAM45A         FANCI         FBXL5 </span>
<span class="co">#&gt;             7             7             2            14             1 </span>
<span class="co">#&gt;         FBXO5          FCAR        FCER1A        FCER1G        FCGR1A </span>
<span class="co">#&gt;             3             2            18             8             7 </span>
<span class="co">#&gt;        FCGR2A         FCGRT          FCN1          FGD4          FGL2 </span>
<span class="co">#&gt;             2             9             8             2             8 </span>
<span class="co">#&gt;           FGR        FKBP1A          FLT3         FOLR3           FOS </span>
<span class="co">#&gt;             9             4            18             7            17 </span>
<span class="co">#&gt;          FOSB         FOSL2          FPR1          FTH1           FTL </span>
<span class="co">#&gt;            17             2             7             8             8 </span>
<span class="co">#&gt;         FTLP3          G0S2     GABARAPL1         GAPDH          GAS6 </span>
<span class="co">#&gt;            11             2             2             4            13 </span>
<span class="co">#&gt;          GAS7           GCA         GINS2        GLIPR1          GLMP </span>
<span class="co">#&gt;             2            12            14             9             3 </span>
<span class="co">#&gt;          GLRX        GLT1D1          GLUL          GMFG         GNAI2 </span>
<span class="co">#&gt;             1             2             2             3             6 </span>
<span class="co">#&gt;          GNAQ           GNS        GPBAR1        GPCPD1          GPX1 </span>
<span class="co">#&gt;             7             2            11             7            12 </span>
<span class="co">#&gt;          GPX4           GRN           GSN         GSTO1         GSTP1 </span>
<span class="co">#&gt;             1            12            18             6             8 </span>
<span class="co">#&gt;          GZMB         H2AFY         H2AFZ         H3F3B         HACD4 </span>
<span class="co">#&gt;            13             6             3            17             2 </span>
<span class="co">#&gt;           HCK         HIPK3      HIST1H4C           HK2      HLA-DPA1 </span>
<span class="co">#&gt;            11             7             5             7            19 </span>
<span class="co">#&gt;      HLA-DPB1      HLA-DQA1       HLA-DRA      HLA-DRB1      HLA-DRB6 </span>
<span class="co">#&gt;            19            19            19            19            19 </span>
<span class="co">#&gt;         HMGB2          HNMT         HSBP1      HSP90AB1         HSPA8 </span>
<span class="co">#&gt;             1             7             7             5            20 </span>
<span class="co">#&gt;           ID1          IER2          IER3          IER5         IFI30 </span>
<span class="co">#&gt;             2            17             2             2            11 </span>
<span class="co">#&gt;        IFITM3        IFNGR1        IFNGR2         IGSF6       IL13RA1 </span>
<span class="co">#&gt;            11             2             2             8             2 </span>
<span class="co">#&gt;          IL1B         IL1RN        IQGAP1         IRAK3       IRF2BP2 </span>
<span class="co">#&gt;            15             2             2             7             2 </span>
<span class="co">#&gt;          IRF7          IRF8          IRS2         ITGAM         ITM2B </span>
<span class="co">#&gt;            13            13             1             7             3 </span>
<span class="co">#&gt;         ITM2C          JAML           JUN          JUNB          JUND </span>
<span class="co">#&gt;            13             9            17            17             7 </span>
<span class="co">#&gt;         KCNE3        KCTD12      KIAA0930        KIF18A         KIF23 </span>
<span class="co">#&gt;             2             9             2             3             3 </span>
<span class="co">#&gt;         KIFC1         KLF10          KLF4          KLF6         KPNA2 </span>
<span class="co">#&gt;            16             2            15            17             3 </span>
<span class="co">#&gt;         LACTB         LAMP2         LAMP5       LAMTOR2       LAMTOR4 </span>
<span class="co">#&gt;             2             2            13             2             7 </span>
<span class="co">#&gt;          LCP1        LGALS1        LGALS2        LGALS3        LILRA2 </span>
<span class="co">#&gt;             6             8             8             8            11 </span>
<span class="co">#&gt;        LILRA4        LILRA5        LILRB2        LILRB3        LILRB4 </span>
<span class="co">#&gt;            13             7            11             2            18 </span>
<span class="co">#&gt;     LINC00877     LINC00996          LMO2        LPCAT2          LRP1 </span>
<span class="co">#&gt;             2            13             7             2             7 </span>
<span class="co">#&gt;        LRRC25        LRRC26         LRRK2          LSM6          LST1 </span>
<span class="co">#&gt;            11            13            15             1             8 </span>
<span class="co">#&gt;         LTA4H          LY96          LYST           LYZ        MAD2L1 </span>
<span class="co">#&gt;             2             7            11             8             3 </span>
<span class="co">#&gt;          MAFB         MAP1A        MARCKS        MCEMP1          MCL1 </span>
<span class="co">#&gt;            11            13            15             7             7 </span>
<span class="co">#&gt;          MCM7         MEGF9        METTL9         MFSD1         MGST1 </span>
<span class="co">#&gt;            14             7             1             2            12 </span>
<span class="co">#&gt;         MGST2          MIDN         MKI67          MNDA         MPEG1 </span>
<span class="co">#&gt;             9             7            16             8             9 </span>
<span class="co">#&gt;        MS4A6A         MS4A7         MSRB1         MYADM         MYBL2 </span>
<span class="co">#&gt;            12            11             7            17            13 </span>
<span class="co">#&gt;          MYL6          MZB1         NAMPT         NAPRT         NAPSB </span>
<span class="co">#&gt;             6            13            15             2            13 </span>
<span class="co">#&gt;          NCF2       NDUFA11        NDUFB1        NDUFS6          NEFH </span>
<span class="co">#&gt;            11             1             2             6             3 </span>
<span class="co">#&gt;         NFAM1          NFE2        NFKBIA        NFKBIZ         NINJ1 </span>
<span class="co">#&gt;             7             2            15             2             1 </span>
<span class="co">#&gt;         NLRP3          NPC2          NPM1         NR4A2          NRG1 </span>
<span class="co">#&gt;             2             8            20            17             2 </span>
<span class="co">#&gt;        NUDT16          NUF2        NUP214        NUSAP1          OAZ1 </span>
<span class="co">#&gt;             2             3            12            16             9 </span>
<span class="co">#&gt;         ODF3B        OGFRL1         OSCAR         OTUD1        P2RY13 </span>
<span class="co">#&gt;            15            15             7             2             2 </span>
<span class="co">#&gt;       PACSIN1          PAK1         PCBP1         PCLAF          PCNA </span>
<span class="co">#&gt;            13             4             1            14            14 </span>
<span class="co">#&gt;         PFDN5          PFN1           PGD          PGLS          PHEX </span>
<span class="co">#&gt;             3             6             7             2            13 </span>
<span class="co">#&gt;        PICALM          PID1         PILRA        PLA2G7         PLAUR </span>
<span class="co">#&gt;             2            15            11             7            15 </span>
<span class="co">#&gt;         PLBD1          PLD4          PLEK          PLP2        PLSCR1 </span>
<span class="co">#&gt;            12            13             9             3             2 </span>
<span class="co">#&gt;        PLXDC2        PLXNB2         PNRC1         POLE4        POLR2L </span>
<span class="co">#&gt;             7             2             3             2             6 </span>
<span class="co">#&gt;          POMP          PPIB      PPP1R14B      PPP1R15A         PRAM1 </span>
<span class="co">#&gt;             6            10            13            17            11 </span>
<span class="co">#&gt;         PRDX3       PRELID1          PROC          PSAP         PSMB3 </span>
<span class="co">#&gt;             1             2            13             8             6 </span>
<span class="co">#&gt;         PTAFR         PTGDS         PTGS2          PTMS         PTPRE </span>
<span class="co">#&gt;             2            13             2            10             9 </span>
<span class="co">#&gt;         PTTG1        PYCARD          PYGL           QKI          QPCT </span>
<span class="co">#&gt;            16             9             2             2             7 </span>
<span class="co">#&gt;     RAB11FIP1         RAB31         RAB32         RAB3D          RAC1 </span>
<span class="co">#&gt;             2             9            15             2             9 </span>
<span class="co">#&gt;        RASSF2        RASSF4         RBM47          RBP7        RCBTB2 </span>
<span class="co">#&gt;             7             7             7            12             1 </span>
<span class="co">#&gt;          RETN          RGS2          RHOA          RHOB          RHOG </span>
<span class="co">#&gt;             7            15             6            15             4 </span>
<span class="co">#&gt;         RIPK2     RN7SL368P     RNA5SP151        RNASE2        RNASE6 </span>
<span class="co">#&gt;             2             2             2             7             9 </span>
<span class="co">#&gt;        RNF130        RNF181        RNF19B      RNU4ATAC      RNVU1-19 </span>
<span class="co">#&gt;             9             1             2             1             2 </span>
<span class="co">#&gt; RP11-1143G9.4  RP11-160E2.6 RP11-363K21.1   RP11-40C6.2 RP11-522I20.3 </span>
<span class="co">#&gt;            12             2             2            20             3 </span>
<span class="co">#&gt;  RP11-700H6.1  RP11-73G16.2        RPL10A        RPL23A        RPL27A </span>
<span class="co">#&gt;             3            13            20            20            20 </span>
<span class="co">#&gt;          RPL3         RPL30         RPL31          RPL5         RPLP0 </span>
<span class="co">#&gt;            20            20            20            20            20 </span>
<span class="co">#&gt;         RPLP2         RPS12         RPS18          RPS2         RPS24 </span>
<span class="co">#&gt;            20            20            20            20             3 </span>
<span class="co">#&gt;         RPS27        RPS27A         RPS29          RPS3         RPS4X </span>
<span class="co">#&gt;            20            20            20            20            20 </span>
<span class="co">#&gt;          RPS5          RPS6          RPSA          RRM2          RTN3 </span>
<span class="co">#&gt;            20            20            20            16             7 </span>
<span class="co">#&gt;          RTN4          RXRA       S100A10       S100A11       S100A12 </span>
<span class="co">#&gt;             6             7             4             8            12 </span>
<span class="co">#&gt;        S100A4        S100A6        S100A8        S100A9        SAMHD1 </span>
<span class="co">#&gt;             8             8            12            12             9 </span>
<span class="co">#&gt;         SAP30          SAT1        SCAMP5        SCPEP1           SCT </span>
<span class="co">#&gt;             3             8            13             2            13 </span>
<span class="co">#&gt;         SDCBP        SEC11A        SEC61B        SEC61G        SEMA4A </span>
<span class="co">#&gt;             2             6            10            10             2 </span>
<span class="co">#&gt;         SERF2         SERP1      SERPINA1      SERPINF1          SGK1 </span>
<span class="co">#&gt;             9             3             8            13            15 </span>
<span class="co">#&gt;          SGO1          SGO2      SH3BGRL3         SIRPA        SIRPB1 </span>
<span class="co">#&gt;             3             3             4             2             7 </span>
<span class="co">#&gt;       SLC11A1       SLC15A3       SLC15A4       SLC16A3       SLC31A2 </span>
<span class="co">#&gt;            11             2            13             2             2 </span>
<span class="co">#&gt;        SLC7A7        SLC8A1         SMCO4        SMIM25         SMIM5 </span>
<span class="co">#&gt;            11             2            11            11            13 </span>
<span class="co">#&gt;         SMPD3          SNCA     SNORD3B-2         SNX10          SOD2 </span>
<span class="co">#&gt;            13             1             2             2            15 </span>
<span class="co">#&gt;         SPCS1         SPG21          SPI1          SRGN        SRPK2P </span>
<span class="co">#&gt;            10             2             8             8             3 </span>
<span class="co">#&gt;         STAB1         STMN1         STX11        STXBP2         SULF2 </span>
<span class="co">#&gt;             7             5             4             8             2 </span>
<span class="co">#&gt;       SULT1A1           SYK        TALDO1          TCF4          TET2 </span>
<span class="co">#&gt;            11             2             7            13             1 </span>
<span class="co">#&gt;          TFEC         TGFBI        TIMM8B         TIMP1         TIMP2 </span>
<span class="co">#&gt;             2             9             6             4             2 </span>
<span class="co">#&gt;           TK1           TKT          TLR2          TLR8       TMEM107 </span>
<span class="co">#&gt;            14            11             2             7             1 </span>
<span class="co">#&gt;      TMEM167A      TMEM170B       TMEM205       TNFAIP2       TNFAIP3 </span>
<span class="co">#&gt;             7             2             2            15            17 </span>
<span class="co">#&gt;      TNFRSF1B      TNFRSF21       TNFSF10      TNFSF13B         TONSL </span>
<span class="co">#&gt;            11            13             2            12             3 </span>
<span class="co">#&gt;         TOP2A          TPI1          TPM2          TPP1          TPX2 </span>
<span class="co">#&gt;            16             4            13             2            16 </span>
<span class="co">#&gt;         TREM1         TRIB1          TSPO        TUBA1B        TUBA1C </span>
<span class="co">#&gt;             2             2             8             3             3 </span>
<span class="co">#&gt;          TUBB        TUBB4B           TXN          TYMP          TYMS </span>
<span class="co">#&gt;             5             3            10             8            16 </span>
<span class="co">#&gt;        TYROBP          UGCG       UNC93B1        UQCR11          USP1 </span>
<span class="co">#&gt;             8            13            18             1             3 </span>
<span class="co">#&gt;         VAMP8          VCAN           VIM          VSIR         VSTM1 </span>
<span class="co">#&gt;             9            12             4             4             7 </span>
<span class="co">#&gt;          WARS          YBX3         YPEL5         YWHAE          ZEB2 </span>
<span class="co">#&gt;            11            15            17             6             9 </span>
<span class="co">#&gt;        ZFAND5         ZFP36       ZFP36L1       ZNF385A         ZNF81 </span>
<span class="co">#&gt;             7            17             2             9             3 </span>
<span class="co">#&gt;        ZSWIM6         ZWINT </span>
<span class="co">#&gt;             2            14</span></code></pre></div>
<p>However, it is much more convenient to look at the correlation heatmaps of these clusters, so let’s generate those:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mcell_plot_gset_cor_mats.html">mcell_plot_gset_cor_mats</a></span>(<span class="dt">gset_id =</span> gset_nm, <span class="dt">scmat_id =</span> sub_mat_id)
<span class="co">#&gt; will downsample the matrix, N= 750</span></code></pre></div>
<p>The plots are generated under a subdirectory of the figures directory named <em>gset_nm</em>.gset_cors. Common practice is to examine them and select the clusters that we want to keep. In our example, In this case we would probably keep clusters 3, 5, 10, 14, 16 and 17 in the gene set: <img src="figs/lateral.gset_cors/3.png" alt="clus3"><img src="figs/lateral.gset_cors/5.png" alt="clus5"><img src="figs/lateral.gset_cors/10.png" alt="clus10"><img src="figs/lateral.gset_cors/14.png" alt="clus14"><img src="figs/lateral.gset_cors/16.png" alt="clus16"><img src="figs/lateral.gset_cors/17.png" alt="clus17"></p>
<p>The rest of the clusters contain mostly genes that are correlated to cell size but not related to the cell cycle.</p>
<p>To create the final filtered gene set, we’ll keep only the clusters we’ve selected in the gene set, and we’ll save the filtered gene set in the database:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mcell_gset_remove_clusts.html">mcell_gset_remove_clusts</a></span>(<span class="dt">gset_id =</span> gset_nm, <span class="dt">filt_clusts =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/c">c</a></span>(<span class="dv">3</span>, <span class="dv">5</span>, <span class="dv">10</span>, <span class="dv">14</span>, <span class="dv">16</span>, <span class="dv">17</span>), <span class="dt">new_id =</span> <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(gset_nm, <span class="st">"_filtered"</span>), <span class="dt">reverse=</span>T)

lateral_gset_id =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/paste">paste0</a></span>(gset_nm, <span class="st">"_filtered"</span>)
lateral_gset =<span class="st"> </span><span class="kw"><a href="../reference/scdb_gset.html">scdb_gset</a></span>(lateral_gset_id)
<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/print">print</a></span>(lateral_gset)
<span class="co">#&gt; An object of class "tgGeneSets"</span>
<span class="co">#&gt; Slot "description":</span>
<span class="co">#&gt; [1] "Cell cycle and stress correlated genes hc K=20 cl_filt"</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "set_names":</span>
<span class="co">#&gt; [1]  3  5 10 14 16 17</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; Slot "gene_set":</span>
<span class="co">#&gt;    AC098614.2         ACTG1        ADGRF3       ALOX5AP     ARHGAP11A </span>
<span class="co">#&gt;             3             5             3            10            14 </span>
<span class="co">#&gt;       ARHGDIB         ASF1B        ATP5G2         ATP5L         AURKA </span>
<span class="co">#&gt;             5            14             3             3             3 </span>
<span class="co">#&gt;         BIRC5      C12orf75         CASZ1         CCNA2         CCNB1 </span>
<span class="co">#&gt;            16            10             3            16            16 </span>
<span class="co">#&gt;         CCNL1          CCR9          CD44         CDCA8        CDKN2C </span>
<span class="co">#&gt;            17             3             3            16             3 </span>
<span class="co">#&gt;         CDKN3          CDT1         CENPF         CENPK         CHEK1 </span>
<span class="co">#&gt;            16            14            16            14            14 </span>
<span class="co">#&gt;         CKAP2         CKS1B          CKS2        COX6B1         COX7B </span>
<span class="co">#&gt;             3             3             3             3             3 </span>
<span class="co">#&gt;         DSCC1         DUSP1           DUT          E2F1          ECT2 </span>
<span class="co">#&gt;             3            17             3            14             3 </span>
<span class="co">#&gt;     EEF1A1P42          EIF1          EZH2         FANCI         FBXO5 </span>
<span class="co">#&gt;             3            17             3            14             3 </span>
<span class="co">#&gt;           FOS          FOSB         GINS2          GLMP          GMFG </span>
<span class="co">#&gt;            17            17            14             3             3 </span>
<span class="co">#&gt;         H2AFZ         H3F3B      HIST1H4C      HSP90AB1          IER2 </span>
<span class="co">#&gt;             3            17             5             5            17 </span>
<span class="co">#&gt;         ITM2B           JUN          JUNB        KIF18A         KIF23 </span>
<span class="co">#&gt;             3            17            17             3             3 </span>
<span class="co">#&gt;         KIFC1          KLF6         KPNA2        MAD2L1          MCM7 </span>
<span class="co">#&gt;            16            17             3             3            14 </span>
<span class="co">#&gt;         MKI67         MYADM          NEFH         NR4A2          NUF2 </span>
<span class="co">#&gt;            16            17             3            17             3 </span>
<span class="co">#&gt;        NUSAP1         PCLAF          PCNA         PFDN5          PLP2 </span>
<span class="co">#&gt;            16            14            14             3             3 </span>
<span class="co">#&gt;         PNRC1          PPIB      PPP1R15A          PTMS         PTTG1 </span>
<span class="co">#&gt;             3            10            17            10            16 </span>
<span class="co">#&gt; RP11-522I20.3  RP11-700H6.1         RPS24          RRM2         SAP30 </span>
<span class="co">#&gt;             3             3             3            16             3 </span>
<span class="co">#&gt;        SEC61B        SEC61G         SERP1          SGO1          SGO2 </span>
<span class="co">#&gt;            10            10             3             3             3 </span>
<span class="co">#&gt;         SPCS1        SRPK2P         STMN1           TK1       TNFAIP3 </span>
<span class="co">#&gt;            10             3             5            14            17 </span>
<span class="co">#&gt;         TONSL         TOP2A          TPX2        TUBA1B        TUBA1C </span>
<span class="co">#&gt;             3            16            16             3             3 </span>
<span class="co">#&gt;          TUBB        TUBB4B           TXN          TYMS          USP1 </span>
<span class="co">#&gt;             5             3            10            16             3 </span>
<span class="co">#&gt;         YPEL5         ZFP36         ZNF81         ZWINT </span>
<span class="co">#&gt;            17            17             3            14</span></code></pre></div>
</div>
<div id="generating-metacells-while-blacklisting-our-lateral-gene-set" class="section level2">
<h2 class="hasAnchor">
<a href="#generating-metacells-while-blacklisting-our-lateral-gene-set" class="anchor"></a>Generating metacells while blacklisting our lateral gene set</h2>
<p>This section will follow the exact same steps performed in the <strong>Running metacell analysis: guided tutorial on 8K PBMCs</strong> vignette in order to generate the K-nn graph, the co-clustering matrix and the metacells, with the small change of disabling the lateral genes from affecting the process.</p>
<p>We start by removing the lateral genes from the list of feature genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">marker_gset =<span class="st"> </span><span class="kw"><a href="../reference/scdb_gset.html">scdb_gset</a></span>(<span class="st">"test_feats"</span>)   
marker_gset =<span class="st"> </span><span class="kw"><a href="../reference/gset_new_restrict_gset.html">gset_new_restrict_gset</a></span>(<span class="dt">gset =</span> marker_gset, <span class="dt">filt_gset =</span> lateral_gset, <span class="dt">inverse =</span> T, <span class="dt">desc =</span> <span class="st">"cgraph gene markers w/o lateral genes"</span>)
<span class="kw"><a href="../reference/scdb_add_gset.html">scdb_add_gset</a></span>(<span class="st">"test_feats_filtered"</span>, marker_gset)</code></pre></div>
<p>We’ll now generate the graph using the filtered list of feature genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mcell_add_cgraph_from_mat_bknn.html">mcell_add_cgraph_from_mat_bknn</a></span>(<span class="dt">mat_id=</span><span class="st">"test"</span>, 
                <span class="dt">gset_id =</span> <span class="st">"test_feats_filtered"</span>, 
                <span class="dt">graph_id=</span><span class="st">"test_graph_filtered"</span>,
                <span class="dt">K=</span><span class="dv">100</span>,
                <span class="dt">dsamp=</span>T)
<span class="co">#&gt; will downsample the matrix, N= 1877</span>
<span class="co">#&gt; will build balanced knn graph on 8276 cells and 893 genes, this can be a bit heavy for &gt;20,000 cells</span>
<span class="co">#&gt; sim graph is missing 39 nodes, out of 8276</span></code></pre></div>
<p>And now simply following the steps of the <strong>Running metacell analysis: guided tutorial on 8K PBMCs</strong> vignette from the K-nn graph until the colored metacell object:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mcell_coclust_from_graph_resamp.html">mcell_coclust_from_graph_resamp</a></span>(
                <span class="dt">coc_id=</span><span class="st">"test_coc500_filtered"</span>, 
                <span class="dt">graph_id=</span><span class="st">"test_graph_filtered"</span>,
                <span class="dt">min_mc_size=</span><span class="dv">20</span>, 
                <span class="dt">p_resamp=</span><span class="fl">0.75</span>, <span class="dt">n_resamp=</span><span class="dv">500</span>)
<span class="co">#&gt; running bootstrap to generate cocluster</span>
<span class="co">#&gt; 0%...2%...3%...4%...6%...7%...9%...10%...11%...13%...15%...16%...17%...18%...20%...21%...22%...23%...25%...26%...28%...30%...31%...32%...34%...35%...37%...38%...39%...40%...42%...44%...45%...47%...49%....51%...52%...53%...55%...56%...58%...59%...60%...62%...63%...65%...66%...67%...69%...70%...72%....74%...75%...77%...78%...79%...81%...82%...84%...85%...99%...100%</span>
<span class="co">#&gt; done resampling</span>

<span class="kw"><a href="../reference/mcell_mc_from_coclust_balanced.html">mcell_mc_from_coclust_balanced</a></span>(
                <span class="dt">coc_id=</span><span class="st">"test_coc500_filtered"</span>, 
                <span class="dt">mat_id=</span> <span class="st">"test"</span>,
                <span class="dt">mc_id=</span> <span class="st">"test_filtered_mc"</span>, 
                <span class="dt">K=</span><span class="dv">30</span>, <span class="dt">min_mc_size=</span><span class="dv">30</span>, <span class="dt">alpha=</span><span class="dv">2</span>)
<span class="co">#&gt; filtered 4999536 left with 741923 based on co-cluster imbalance</span>
<span class="co">#&gt; building metacell object, #mc 77</span>
<span class="co">#&gt; add batch counts</span>
<span class="co">#&gt; compute footprints</span>
<span class="co">#&gt; compute absolute ps</span>
<span class="co">#&gt; compute coverage ps</span>
<span class="co">#&gt; reordering metacells by hclust and most variable two markers</span>
<span class="co">#&gt; reorder on HLA-DRB1 vs LDHB</span>

<span class="kw"><a href="../reference/mcell_mc_split_filt.html">mcell_mc_split_filt</a></span>(<span class="dt">new_mc_id=</span><span class="st">"test_filtered_mc_f"</span>, 
            <span class="dt">mc_id=</span><span class="st">"test_filtered_mc"</span>, 
            <span class="dt">mat_id=</span><span class="st">"test"</span>,
            <span class="dt">T_lfc=</span><span class="dv">3</span>, <span class="dt">plot_mats=</span>F)
<span class="co">#&gt; starting split outliers</span>
<span class="co">#&gt; splitting metacell 2</span>
<span class="co">#&gt; add batch counts</span>
<span class="co">#&gt; compute footprints</span>
<span class="co">#&gt; compute absolute ps</span>
<span class="co">#&gt; compute coverage ps</span>

marks_colors =<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/read.table">read.table</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/system.file">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"pbmc_mc_colorize.txt"</span>, <span class="dt">package=</span><span class="st">"metacell"</span>), <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">h=</span>T, <span class="dt">stringsAsFactors=</span>F)
<span class="kw"><a href="../reference/mc_colorize.html">mc_colorize</a></span>(<span class="st">"test_filtered_mc_f"</span>, <span class="dt">marker_colors=</span>marks_colors)</code></pre></div>
<p>It is convenient not to restrict the selection of genes that will be displayed in the heatmap of genes and metacells, but to mark the lateral genes that did not affect the partitioning of cells to metacells:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mcell_gset_from_mc_markers.html">mcell_gset_from_mc_markers</a></span>(<span class="dt">gset_id=</span><span class="st">"test_filtered_markers"</span>, <span class="dt">mc_id=</span><span class="st">"test_filtered_mc_f"</span>, <span class="dt">blacklist_gset_id=</span>lateral_gset_id)
<span class="kw"><a href="../reference/mcell_gset_from_mc_markers.html">mcell_gset_from_mc_markers</a></span>(<span class="dt">gset_id=</span><span class="st">"test_filtered_markers_lat"</span>, <span class="dt">mc_id=</span><span class="st">"test_filtered_mc_f"</span>, <span class="dt">filt_gset_id=</span>lateral_gset_id)
    
<span class="kw"><a href="../reference/mcell_mc_plot_marks.html">mcell_mc_plot_marks</a></span>(<span class="dt">mc_id=</span><span class="st">"test_filtered_mc_f"</span>, <span class="dt">gset_id=</span><span class="st">"test_filtered_markers"</span>, <span class="dt">mat_id=</span><span class="st">"test"</span>, <span class="dt">lateral_gset=</span><span class="st">"test_filtered_markers_lat"</span>)</code></pre></div>
<p><img src="figs/test_filtered_mc_f.cells_heat_marks.png" alt="heatmap_marks"> The marker genes in the <em>test_filtered_markers</em> gene set do not contain any lateral gene (shown in black in the heatmap), and the <em>test_filtered_markers_lat</em> gene set contain the lateral genes that were selected as markers (shown in red).</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#setting-up">Setting up</a></li>
      <li><a href="#creating-the-gene-correlation-matrix">Creating the gene correlation matrix</a></li>
      <li><a href="#fine-clustering-and-filtering-of-the-gene-set">Fine clustering and filtering of the gene set</a></li>
      <li><a href="#generating-metacells-while-blacklisting-our-lateral-gene-set">Generating metacells while blacklisting our lateral gene set</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by <a href="http://compgenomics.weizmann.ac.il/tanay/">Amos Tanay</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  
<script src="https://cdnjs.cloudflare.com/ajax/libs/docsearch.js/2.6.1/docsearch.min.js" integrity="sha256-GKvGqXDznoRYHCwKXGnuchvKSwmx9SRMrZOTh2g4Sb0=" crossorigin="anonymous"></script><script>
  docsearch({
    
    
    apiKey: 'ca4e16717bb635cc1f804f6a25717f9b',
    indexName: 'metacell',
    inputSelector: 'input#search-input.form-control',
    transformData: function(hits) {
      return hits.map(function (hit) {
        hit.url = updateHitURL(hit);
        return hit;
      });
    }
  });
</script>
</body>
</html>
