<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Annotation of metacells • metacell</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/cosmo/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/all.min.css" integrity="sha256-nAmazAk6vS34Xqo0BSrTb+abbtFlgsFK7NKSi6o7Y78=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.7.1/css/v4-shims.min.css" integrity="sha256-6qHlizsOWFskGlwVOKuns+D1nB6ssZrHQrNj1wGplHc=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/headroom.min.js" integrity="sha256-DJFC1kqIhelURkuza0AvYal5RxMtpzLjFhsnVIeuk+U=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.9.4/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Annotation of metacells">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">metacell</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">0.3.41</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../index.html">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../articles/a-basic_pbmc8k.html">Usage</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li class="dropdown-header">Core</li>
    <li>
      <a href="../articles/a-basic_pbmc8k.html">Running metacell analysis: guided tutorial on 8K PBMCs</a>
    </li>
    <li>
      <a href="../articles/b-geneset_by_anchor_pbmc8k.html">Supervised filtering of feature genes: guided tutorial on 8K PBMCs</a>
    </li>
    <li>
      <a href="../articles/c-metacell_annotation.html">Annotation of metacells</a>
    </li>
    <li class="divider">
    <li class="dropdown-header">Misc</li>
    <li>
      <a href="../articles/d-amphimedon.html">Analyzing whole-organism scRNA-seq data with metacell</a>
    </li>
    <li>
      <a href="../articles/e-data_convert.html">UMI matrix conversion to standard formats</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a></a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>Annotation of metacells</h1>
                        <h4 class="author">Yaniv Lubling</h4>
            
            <h4 class="date">2018-12-26</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/tanaylab/metacell/blob/master/vignettes/c-metacell_annotation.Rmd"><code>vignettes/c-metacell_annotation.Rmd</code></a></small>
      <div class="hidden name"><code>c-metacell_annotation.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>This vignette demonstrates several ways to annotate metacells. It uses the genes relative enrichment of the metacells and the cross-metacell similarity matrix (confusion matrix) to identify metacells of a certain cell-type or cell state. The dataset used in this vignette is single-cell RNA-seq of T cells from lung cancer patients, from the <a href="https://www.ncbi.nlm.nih.gov/pubmed/29942094">Guo et al, Nat Med. 2018</a> paper.</p>
<p>Annotation of metacells is done by assigning a color to each metacell in the <span class="citation">@colors</span> vector in the metacell object. Information on the colors is stored in the <span class="citation">@color_key</span> data frame. It contains the columns ‘group’ (name of the type/state assigned to the metacell) and ‘color’, and might contain additional columns, depending on the method used to annotate the metacells, as will be shown below.</p>
</div>
<div id="setting-up" class="section level2">
<h2 class="hasAnchor">
<a href="#setting-up" class="anchor"></a>Setting up</h2>
<p>The metacells partitioning we use here was generated from the transcript per million reads (TPM) input table Guo et al. supply. For more details see the guo2018.r script at <a href="https://bitbucket.org/tanaylab/li_et_al_cell_2018_melanoma_scrna/src/default/">Li_et_al_Cell_2018_Melanoma_scRNA</a> code repository. We’ll first download the input objects, specifically the unannotated metacell object. #<code>{r, eval=TRUE, warning=FALSE} #download.file("http://www.wisdom.weizmann.ac.il/~lubling/metac_data/metacell_annotation/lung_db.tar.gz", destfile = #"lung_db.tar.gz") #system("tar xfz lung_db.tar.gz") #file.remove("lung_db.tar.gz") #</code></p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span>(<span class="st">"lung_db"</span>, <span class="dt">showWarnings =</span> F)
<span class="cf">for</span> (f <span class="cf">in</span> <span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">"cgraph.guo2018_tpm_scaled_filt_Tumor_Normal_Blood.Rda"</span>,
            <span class="st">"coclust.guo2018_tpm_scaled_filt_Tumor_Normal_Blood.Rda"</span>,
            <span class="st">"gset.guo2018_lateral.Rda"</span>,
            <span class="st">"gstat.guo2018_tpm_scaled_filt_Tumor_Normal_Blood.Rda"</span>,
            <span class="st">"lfp_screenshot.png"</span>,
            <span class="st">"mat.guo2018_tpm_scaled_filt_guo2018_lateral.Rda"</span>,
            <span class="st">"mat.guo2018_tpm_scaled_filt_Tumor_Normal_Blood.Rda"</span>,
            <span class="st">"mc.guo2018_tpm_scaled_filt_Tumor_Normal_Blood_outClean_nonAnn.Rda"</span>)) {
    <span class="kw"><a href="https://rdrr.io/r/utils/download.file.html">download.file</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"http://www.wisdom.weizmann.ac.il/~lubling/metac_data/metacell_annotation/"</span>, f), <span class="dt">destfile=</span><span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(<span class="st">"lung_db/"</span>, f))
}</code></pre></div>
<p>Now let’s load the metacell package and define the output directories:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(metacell)
<span class="kw"><a href="https://rdrr.io/r/base/library.html">library</a></span>(dplyr)
<span class="co">#&gt; </span>
<span class="co">#&gt; Attaching package: 'dplyr'</span>
<span class="co">#&gt; The following objects are masked from 'package:stats':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     filter, lag</span>
<span class="co">#&gt; The following objects are masked from 'package:base':</span>
<span class="co">#&gt; </span>
<span class="co">#&gt;     intersect, setdiff, setequal, union</span>

<span class="kw"><a href="../reference/scdb_init.html">scdb_init</a></span>(<span class="st">"lung_db"</span>, <span class="dt">force_reinit=</span>T)
<span class="co">#&gt; initializing scdb to lung_db</span>

<span class="kw"><a href="https://rdrr.io/r/base/files2.html">dir.create</a></span>(<span class="st">"lung_figs"</span>, <span class="dt">showWarnings =</span> F)
<span class="kw"><a href="../reference/scfigs_init.html">scfigs_init</a></span>(<span class="st">"lung_figs"</span>)</code></pre></div>
<p>For the sake of code clarity, lets define variables for the main metacell object IDs we’ll soon use:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mc_non_ann_id   =<span class="st"> "guo2018_tpm_scaled_filt_Tumor_Normal_Blood_outClean_nonAnn"</span>
mc_id           =<span class="st"> "guo2018_tpm_scaled_filt_Tumor_Normal_Blood_outClean"</span>
graph_id        =<span class="st"> "guo2018_tpm_scaled_filt_Tumor_Normal_Blood"</span>
mat_id          =<span class="st"> </span>graph_id
lateral_gset_id =<span class="st"> "guo2018_lateral"</span></code></pre></div>
</div>
<div id="initial-exploration-of-the-metacells" class="section level2">
<h2 class="hasAnchor">
<a href="#initial-exploration-of-the-metacells" class="anchor"></a>Initial exploration of the metacells</h2>
<p>It is often useful to start with semi-random coloring of metacells by assignment of sequentail colors to metacells. Since metacells are ordered by default, similar metacells will tend to have similar colors by this process. We’ll first make a copy of the non-annotated input metacell object, in order not to overwrite it (we’ll need it later on). Coloring is done by <em>mc_colorize default</em> function. The user can override the default color specturm used by supplying her own color spectrum (typically the output of <em>colorRampPalette</em>) to the <em>spectrum</em> parameter.</p>
<p>The function assigned colors in the <span class="citation">@colors</span> slot, yet the <span class="citation">@color_key</span> data frame is empty, since these colors are arbitrary and are not associated with any annotation.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mc_non_ann =<span class="st"> </span><span class="kw"><a href="../reference/scdb_mc.html">scdb_mc</a></span>(mc_non_ann_id)
<span class="kw"><a href="../reference/scdb_add_mc.html">scdb_add_mc</a></span>(mc_id, mc_non_ann)

<span class="kw"><a href="../reference/mc_colorize_default.html">mc_colorize_default</a></span>(mc_id)
mc =<span class="st"> </span><span class="kw"><a href="../reference/scdb_mc.html">scdb_mc</a></span>(mc_id)

<span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(mc<span class="op">@</span>colors)
<span class="co">#&gt;  [1] "#FFFFFF" "#F9F9F9" "#F3F3F3" "#EDEDED" "#E7E7E7" "#E1E1E1" "#DBDBDB"</span>
<span class="co">#&gt;  [8] "#D5D5D5" "#CFCFCF" "#CACACA" "#C4C4C4" "#BEBEBE" "#B9B9B9" "#B3B3B3"</span>
<span class="co">#&gt; [15] "#AEAEAE" "#AAA9A8" "#B5AFA6" "#C1B4A5" "#CCBAA3" "#D8C0A1" "#E4C59F"</span>
<span class="co">#&gt; [22] "#EFCB9D" "#FBD19B" "#F4C68E" "#E5B37C" "#D5A06A" "#C58D58" "#B67945"</span>
<span class="co">#&gt; [29] "#A66633" "#975321" "#8E4712" "#9E540F" "#AD610D" "#BD6E0A" "#CC7B08"</span>
<span class="co">#&gt; [36] "#DC8805" "#EB9503" "#FBA200" "#FF9300" "#FF7D00" "#FF6700" "#FF5100"</span>
<span class="co">#&gt; [43] "#FF3B00" "#FF2400" "#FF0E00" "#FA010A" "#ED052A" "#E10A4B" "#D40E6B"</span>
<span class="co">#&gt; [50] "#C7128B" "#BA16AB" "#AE1BCC" "#A11FEC" "#8C1CF1" "#7717F3" "#6113F5"</span>
<span class="co">#&gt; [57] "#4C0FF7" "#360AF9" "#2106FB" "#0B02FD" "#000FFF" "#0031FF" "#0053FF"</span>
<span class="co">#&gt; [64] "#0075FF" "#0098FF" "#00BAFF" "#00DCFF" "#00FFFF"</span>
<span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(mc<span class="op">@</span>color_key)
<span class="co">#&gt; data frame with 0 columns and 0 rows</span></code></pre></div>
<p>A concevenient way to explore the metacells is to generate a heatmap of gene enrichments on single cells or metacells. The input data is the genes relative footprint values across metacells over the median metacell footprint (the table under the <span class="citation">@mc_fp</span> slot). We’ll usually use the log2 value of these ratios, and will term the matrix of log ratios <em>lfp</em> in this vignette.</p>
<p>We’ll first select which genes to present with <em>mcell_gset_from_mc_markers</em>. Its basic usage expects an input metacell id (mc_id) and the name of the gene set to store the genes in (gset_id). This function screens for highly varying genes (max absolute lfp value above the config parameter <em>scm_mc_mark_min_gene_fold</em>, 1.5 by default) and selects the top N varying genes per metacell (N is defined in the <em>scm_mc_mark_k_per_clust</em> parameter, 5 by default). The union of the genes selected per metacell defines the output gene set.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mcell_gset_from_mc_markers.html">mcell_gset_from_mc_markers</a></span>(<span class="dt">gset_id =</span> mc_id, <span class="dt">mc_id =</span> mc_id)</code></pre></div>
<p>Now we’ll generate the heatmap, supplying the metacell id and the gene set id of the genes to show, and since we plot the cells (and not the metacells), also the mat id (from which the raw UMI counts are taken). The main configuration parameters that control the heatmap are <em>mcp_heatmap_height</em> and <em>mcp_heatmap_width</em> (figure height and width in pixels), <em>mcp_heatmap_text_cex</em> (text size, passed as cex param to <em>text</em> function) and <em>mcp_heatmap_alt_side_text</em> (logical, if true, plot odd gene names to the right and even to the left of the heatmap).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mcell_mc_plot_marks.html">mcell_mc_plot_marks</a></span>(<span class="dt">mc_id =</span> mc_id, <span class="dt">gset_id =</span> mc_id, <span class="dt">mat_id =</span> mat_id, <span class="dt">plot_cells=</span>T)</code></pre></div>
<div class="figure">
<img src="lung_figs/guo2018_tpm_scaled_filt_Tumor_Normal_Blood_outClean.cells_heat_marks.png" alt="Varying genes footprint on single-cells, ordered by metacells" width="100%"><p class="caption">
Varying genes footprint on single-cells, ordered by metacells
</p>
</div>
<p>Lateral genes are enriched in several metacells regardless of the cell type or state. The obious example for such lateral gene program is cell cycle related genes. We often filter lateral genes from being features that affect the creation of metacells. When plotting gene-metacell heatmap, we can filter lateral genes from the displayed genes by supplying the lateral gene set id to the <em>blacklist_gset_id</em> parameter in <em>mcell_gset_from_mc_markers</em>. However, we might still want to see the expression of these genes across metacells in the heatmap, yet without letting them affect the ordering of metacells in the heatmap. To do so we generate 2 gene sets - both of highly variable genes across metacells, the first without and the second with only lateral genes. The first will be supplied to the gset_id param in <em>mcell_mc_plot_marks</em> and the second to the <em>lateral_gset_id</em> param. In this data, the lateral gene set contains cell cycle, IFNa response and stress related genes. The lateral genes are marked in red and are shown on at the top of the heatmap.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="co"># prevent the lateral genes from being in the gene set</span>
<span class="kw"><a href="../reference/mcell_gset_from_mc_markers.html">mcell_gset_from_mc_markers</a></span>(<span class="dt">gset_id =</span> mc_id, <span class="dt">mc_id =</span> mc_id, <span class="dt">blacklist_gset_id =</span> lateral_gset_id)

<span class="co"># select variable genes from the lateral gene set</span>
<span class="kw"><a href="../reference/mcell_gset_from_mc_markers.html">mcell_gset_from_mc_markers</a></span>(<span class="dt">gset_id =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(mc_id, <span class="st">"_lateral"</span>), <span class="dt">mc_id =</span> mc_id, <span class="dt">filt_gset_id =</span> lateral_gset_id)
        
<span class="kw"><a href="../reference/mcell_mc_plot_marks.html">mcell_mc_plot_marks</a></span>(<span class="dt">mc_id =</span> mc_id, <span class="dt">gset_id =</span> mc_id, <span class="dt">mat_id =</span> mat_id, <span class="dt">fig_fn =</span> <span class="kw"><a href="../reference/scfigs_fn.html">scfigs_fn</a></span>(mc_id, <span class="st">"cells_heat_marks_lat"</span>), <span class="dt">lateral_gset_id =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(mc_id, <span class="st">"_lateral"</span>), <span class="dt">plot_cells=</span>T)</code></pre></div>
<div class="figure">
<img src="lung_figs/guo2018_tpm_scaled_filt_Tumor_Normal_Blood_outClean.cells_heat_marks_lat.png" alt="Varying genes footrprint on single-cells, ordered by metacells, lateral genes on top in red" width="100%"><p class="caption">
Varying genes footrprint on single-cells, ordered by metacells, lateral genes on top in red
</p>
</div>
<p>Another quick and popular way to get a feeling on metacell similarity (and grouping) is to generate a 2D projection of cells and metacells, where the coordinates of each metacell and the edges between metacells are dictated by inter-metacell similarity. <em>mcell_mc2d_force_knn</em> builds the metacell 2D projection (mc2d object) and <em>mcell_mc2d_plot</em> generates the plot. The relevant configuration parameters affecting the plot are <em>mcell_mc2d_height</em> and <em>mcell_mc2d_width</em> (dimension in pixels), <em>mcell_mc2d_plot_key</em> (logical, whether to show the legend), <em>mcell_mc2d_cex</em> and <em>mcell_mc2d_legend_cex</em> (cells and legend size).</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mcell_mc2d_force_knn.html">mcell_mc2d_force_knn</a></span>(<span class="dt">mc2d_id =</span> mc_id, <span class="dt">mc_id =</span> mc_id, <span class="dt">graph_id =</span> graph_id)
<span class="co">#&gt; comp mc graph using the graph guo2018_tpm_scaled_filt_Tumor_Normal_Blood and K 20</span>
<span class="kw"><a href="../reference/mcell_mc2d_plot.html">mcell_mc2d_plot</a></span>(<span class="dt">mc2d_id =</span> mc_id, <span class="dt">plot_edges =</span> <span class="ot">TRUE</span>) 
<span class="co">#&gt; png </span>
<span class="co">#&gt;   2</span></code></pre></div>
<pre><code>#&gt; [1] TRUE</code></pre>
<div class="figure">
<img src="lung_figs/guo2018_tpm_scaled_filt_Tumor_Normal_Blood_outClean.2d_graph_proj_default_col.png" alt="Single-cell and metacells 2D projection" width="100%"><p class="caption">
Single-cell and metacells 2D projection
</p>
</div>
</div>
<div id="metacells-gene-enrichment-table" class="section level2">
<h2 class="hasAnchor">
<a href="#metacells-gene-enrichment-table" class="anchor"></a>Metacells gene enrichment table</h2>
<p>As discussed above, the central way to characterize metacells is via the metacell gene enrichment table (termed here <em>lfp</em>, values in log2 ratio over median gene expression across metacells). We can generate a heatmap of the lfp, marking lateral genes in red and disabling them from affecting the metacells order (columns) as we did before when plotting a heatmap of genes and single-cells:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mcell_mc_plot_marks.html">mcell_mc_plot_marks</a></span>(<span class="dt">mc_id =</span> mc_id, <span class="dt">gset_id =</span> mc_id, <span class="dt">mat_id =</span> mat_id, <span class="dt">fig_fn =</span> <span class="kw"><a href="../reference/scfigs_fn.html">scfigs_fn</a></span>(mc_id, <span class="st">"mc_heat_marks_lat"</span>), <span class="dt">lateral_gset_id =</span> <span class="kw"><a href="https://rdrr.io/r/base/paste.html">paste0</a></span>(mc_id, <span class="st">"_lateral"</span>), <span class="dt">plot_cells=</span>F)</code></pre></div>
<div class="figure">
<img src="lung_figs/guo2018_tpm_scaled_filt_Tumor_Normal_Blood_outClean.mc_heat_marks_lat.png" alt="Varying genes footrprint on metacells, ordered by metecalls, blacklisted genes on top in red" width="100%"><p class="caption">
Varying genes footrprint on metacells, ordered by metecalls, blacklisted genes on top in red
</p>
</div>
<p>It is also useful to do scatter plots for specific genes. In the example below, we’re interested if the data contain Treg cells, so we plot CTLA4 vs CD4, and indeed there is a group of metacells enriched in both genes.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">lfp =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(mc<span class="op">@</span>mc_fp)

plt =<span class="st"> </span><span class="cf">function</span>(gene1, gene2, lfp, colors) 
{
    <span class="kw"><a href="https://rdrr.io/r/graphics/plot.html">plot</a></span>(lfp[gene1, ], lfp[gene2, ], <span class="dt">pch=</span><span class="dv">21</span>, <span class="dt">cex=</span><span class="dv">3</span>, <span class="dt">bg=</span>colors, <span class="dt">xlab=</span>gene1, <span class="dt">ylab=</span>gene2)
    <span class="kw"><a href="https://rdrr.io/r/graphics/text.html">text</a></span>(lfp[gene1, ], lfp[gene2, ], <span class="kw"><a href="https://rdrr.io/r/base/colnames.html">colnames</a></span>(lfp))
    
}
<span class="kw">plt</span>(<span class="dt">gene1 =</span> <span class="st">'CD4'</span>, <span class="dt">gene2 =</span> <span class="st">'CTLA4'</span>, <span class="dt">lfp =</span> lfp, <span class="dt">colors =</span> mc<span class="op">@</span>colors)</code></pre></div>
<p><img src="c-metacell_annotation_files/figure-html/unnamed-chunk-14-1.png" width="480"></p>
<p>It is also useful to browse the lfp table itself. <em>mcell_mc_export_tab</em> exports the lfp table to a tab-delimited file. It filters genes with maximal lfp value above the <em>T_fold</em> parameter. The function reports the metacell size (n_cells) and mean UMIs per metacell (mean_umis), which is a proxy for the cell size. The group field contains the annotation of the metacell (this will be informative after the metacell coloring functions we’ll present below). The metadata field names (columns in the mat object <span class="citation">@cell_metadata</span> table) can be supplied to the <em>metadata_fields</em> parameter. The function would then breakdown cells in each metacell on the values of the metadata field. Very useful to characterize the metacells if our data has informative metadata features.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mcell_mc_export_tab.html">mcell_mc_export_tab</a></span>(<span class="dt">mc_id =</span> mc_id, <span class="dt">gstat_id =</span> mat_id, <span class="dt">mat_id =</span> mat_id, <span class="dt">T_fold=</span><span class="dv">2</span>, <span class="dt">metadata_fields=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'Patient'</span>, <span class="st">'CD8_gate'</span>, <span class="st">'Sex'</span>, <span class="st">'Stage'</span>, <span class="st">'Histology'</span>))
lfp_tab =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="kw"><a href="../reference/scfigs_fn.html">scfigs_fn</a></span>(mc_id, <span class="st">"log2_mc_fp"</span>, <span class="dt">ext =</span> <span class="st">"txt"</span>), <span class="dt">header=</span>F, <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">stringsAsFactors =</span> F)
knitr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(lfp_tab[<span class="dv">1</span><span class="op">:</span><span class="dv">40</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">6</span>])</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">V1</th>
<th align="left">V2</th>
<th align="right">V3</th>
<th align="right">V4</th>
<th align="right">V5</th>
<th align="right">V6</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Histology</td>
<td align="left">adenocarcinoma</td>
<td align="right">160.00</td>
<td align="right">217.00</td>
<td align="right">135.00</td>
<td align="right">201.00</td>
</tr>
<tr class="even">
<td align="left">Histology</td>
<td align="left">squamous cell carcinoma</td>
<td align="right">7.00</td>
<td align="right">27.00</td>
<td align="right">2.00</td>
<td align="right">5.00</td>
</tr>
<tr class="odd">
<td align="left">Stage</td>
<td align="left">I</td>
<td align="right">137.00</td>
<td align="right">193.00</td>
<td align="right">37.00</td>
<td align="right">14.00</td>
</tr>
<tr class="even">
<td align="left">Stage</td>
<td align="left">III</td>
<td align="right">27.00</td>
<td align="right">48.00</td>
<td align="right">99.00</td>
<td align="right">192.00</td>
</tr>
<tr class="odd">
<td align="left">Stage</td>
<td align="left">IV</td>
<td align="right">3.00</td>
<td align="right">3.00</td>
<td align="right">1.00</td>
<td align="right">0.00</td>
</tr>
<tr class="even">
<td align="left">Sex</td>
<td align="left">F</td>
<td align="right">139.00</td>
<td align="right">192.00</td>
<td align="right">39.00</td>
<td align="right">188.00</td>
</tr>
<tr class="odd">
<td align="left">Sex</td>
<td align="left">M</td>
<td align="right">28.00</td>
<td align="right">52.00</td>
<td align="right">98.00</td>
<td align="right">18.00</td>
</tr>
<tr class="even">
<td align="left">CD8_gate</td>
<td align="left">-</td>
<td align="right">166.00</td>
<td align="right">243.00</td>
<td align="right">129.00</td>
<td align="right">1.00</td>
</tr>
<tr class="odd">
<td align="left">CD8_gate</td>
<td align="left">+</td>
<td align="right">1.00</td>
<td align="right">1.00</td>
<td align="right">8.00</td>
<td align="right">205.00</td>
</tr>
<tr class="even">
<td align="left">Patient</td>
<td align="left">P0616A</td>
<td align="right">13.00</td>
<td align="right">10.00</td>
<td align="right">15.00</td>
<td align="right">0.00</td>
</tr>
<tr class="odd">
<td align="left">Patient</td>
<td align="left">P0616P</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">2.00</td>
</tr>
<tr class="even">
<td align="left">Patient</td>
<td align="left">P0617</td>
<td align="right">21.00</td>
<td align="right">21.00</td>
<td align="right">91.00</td>
<td align="right">7.00</td>
</tr>
<tr class="odd">
<td align="left">Patient</td>
<td align="left">P0619</td>
<td align="right">2.00</td>
<td align="right">4.00</td>
<td align="right">6.00</td>
<td align="right">3.00</td>
</tr>
<tr class="even">
<td align="left">Patient</td>
<td align="left">P0706</td>
<td align="right">0.00</td>
<td align="right">1.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
</tr>
<tr class="odd">
<td align="left">Patient</td>
<td align="left">P0729</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">1.00</td>
<td align="right">177.00</td>
</tr>
<tr class="even">
<td align="left">Patient</td>
<td align="left">P0913</td>
<td align="right">4.00</td>
<td align="right">23.00</td>
<td align="right">1.00</td>
<td align="right">5.00</td>
</tr>
<tr class="odd">
<td align="left">Patient</td>
<td align="left">P1010</td>
<td align="right">6.00</td>
<td align="right">161.00</td>
<td align="right">4.00</td>
<td align="right">1.00</td>
</tr>
<tr class="even">
<td align="left">Patient</td>
<td align="left">P1011</td>
<td align="right">3.00</td>
<td align="right">3.00</td>
<td align="right">1.00</td>
<td align="right">0.00</td>
</tr>
<tr class="odd">
<td align="left">Patient</td>
<td align="left">P1118</td>
<td align="right">9.00</td>
<td align="right">9.00</td>
<td align="right">0.00</td>
<td align="right">3.00</td>
</tr>
<tr class="even">
<td align="left">Patient</td>
<td align="left">P1120</td>
<td align="right">1.00</td>
<td align="right">3.00</td>
<td align="right">0.00</td>
<td align="right">1.00</td>
</tr>
<tr class="odd">
<td align="left">Patient</td>
<td align="left">P1202</td>
<td align="right">108.00</td>
<td align="right">9.00</td>
<td align="right">0.00</td>
<td align="right">1.00</td>
</tr>
<tr class="even">
<td align="left">Patient</td>
<td align="left">P1208</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">6.00</td>
</tr>
<tr class="odd">
<td align="left">Patient</td>
<td align="left">P1219</td>
<td align="right">0.00</td>
<td align="right">0.00</td>
<td align="right">18.00</td>
<td align="right">0.00</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">mean_umis</td>
<td align="right">19930.12</td>
<td align="right">19913.84</td>
<td align="right">19939.01</td>
<td align="right">19951.63</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">n_cells</td>
<td align="right">167.00</td>
<td align="right">244.00</td>
<td align="right">137.00</td>
<td align="right">206.00</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">group</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
<td align="right">NA</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">mc_id</td>
<td align="right">1.00</td>
<td align="right">2.00</td>
<td align="right">3.00</td>
<td align="right">4.00</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">TANK</td>
<td align="right">-0.15</td>
<td align="right">-0.12</td>
<td align="right">-0.20</td>
<td align="right">0.10</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">LINC00892</td>
<td align="right">0.50</td>
<td align="right">0.47</td>
<td align="right">0.36</td>
<td align="right">0.19</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">OST4</td>
<td align="right">-0.56</td>
<td align="right">0.26</td>
<td align="right">0.25</td>
<td align="right">0.30</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">LINC00861</td>
<td align="right">0.05</td>
<td align="right">0.28</td>
<td align="right">0.45</td>
<td align="right">-0.32</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">MIR1292</td>
<td align="right">0.28</td>
<td align="right">0.10</td>
<td align="right">-0.03</td>
<td align="right">-0.40</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">MIR1282</td>
<td align="right">-0.81</td>
<td align="right">0.16</td>
<td align="right">0.13</td>
<td align="right">0.19</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">PCNA-AS1</td>
<td align="right">0.02</td>
<td align="right">0.10</td>
<td align="right">-0.22</td>
<td align="right">-0.02</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">ZEB2-AS1</td>
<td align="right">0.91</td>
<td align="right">0.70</td>
<td align="right">1.00</td>
<td align="right">0.05</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">MIR1244-2</td>
<td align="right">-0.13</td>
<td align="right">-0.26</td>
<td align="right">-0.70</td>
<td align="right">0.07</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">MIR3917</td>
<td align="right">0.18</td>
<td align="right">0.14</td>
<td align="right">0.07</td>
<td align="right">0.00</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">MIR3656</td>
<td align="right">0.21</td>
<td align="right">0.07</td>
<td align="right">-0.20</td>
<td align="right">-0.01</td>
</tr>
<tr class="odd">
<td align="left"></td>
<td align="left">MIR3652</td>
<td align="right">0.34</td>
<td align="right">0.32</td>
<td align="right">-0.34</td>
<td align="right">-0.39</td>
</tr>
<tr class="even">
<td align="left"></td>
<td align="left">MIR3615</td>
<td align="right">1.19</td>
<td align="right">0.83</td>
<td align="right">1.14</td>
<td align="right">-0.83</td>
</tr>
</tbody>
</table>
It is convenient to browse this table with MS Excel or a similar tool, and to color the gene’s lfp values and the different metadata header rows (seperately, because of their different range), using the ‘condistional formatting’ options. The screenshot below shows a formatted lfp table. It is immediate to classify metacells as CD4 or CD8, to see patient specific metacells or pan-patient metacells etc. It is also useful to freeze the header rows and the gene names and then to sort the table content according to the metacell we’re interseted in (in the screenshot below it’s metacell #23), to quickly see its enriched genes. Since the metacells are clustered, it also makes the group of similar metacells pop-up very clearly.
<div class="figure">
<img src="lung_db/lfp_screenshot.png" alt="Example of formatted lfp table" width="100%"><p class="caption">
Example of formatted lfp table
</p>
</div>
</div>
<div id="manual-annotation-by-thresholding-gene-enrichment" class="section level2">
<h2 class="hasAnchor">
<a href="#manual-annotation-by-thresholding-gene-enrichment" class="anchor"></a>Manual annotation by thresholding gene enrichment</h2>
<p>In many cases we’ll have some prior knowledge on the cell types we expect to be represented by the metacells. Scatter plots of lfp values of selected genes can highlight these metacells. For instance, the plots below show metacells enriched in IL2RA and FOXP3, which are probably Tregs, CD4+ metacells that strongly express CXCL13, which probably represent Tfh cells, a single proliferating metacell enriched with TOP2A, and naive T cells, expressiong IL7R and TCF7. We can set a threshold on the lfp value of these genes (marked with dashed line) to assign an annotation to these metacells:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">genes1 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'IL2RA'</span>, <span class="st">'CD4'</span>, <span class="st">'IL7R'</span>, <span class="st">'CD8A'</span>)
genes2 =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="st">'FOXP3'</span>, <span class="st">'CXCL13'</span>, <span class="st">'TCF7'</span>, <span class="st">'TOP2A'</span>)
cutoffs =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/base/Log.html">log2</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="fl">2.25</span>, <span class="dv">16</span>, <span class="dv">2</span>, <span class="dv">2</span>))
<span class="kw"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="dt">mfrow=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">2</span>,<span class="dv">2</span>))
<span class="kw"><a href="https://rdrr.io/r/graphics/par.html">par</a></span>(<span class="dt">mar=</span><span class="kw"><a href="https://rdrr.io/r/base/c.html">c</a></span>(<span class="dv">4</span>,<span class="dv">4</span>,<span class="dv">1</span>,<span class="dv">1</span>))
<span class="cf">for</span> (i <span class="cf">in</span> <span class="kw"><a href="https://rdrr.io/r/base/seq.html">seq_along</a></span>(genes1)) {
    <span class="kw">plt</span>(<span class="dt">gene1 =</span> genes1[i], <span class="dt">gene2 =</span> genes2[i], <span class="dt">lfp =</span> lfp, <span class="dt">colors =</span> mc<span class="op">@</span>colors)
    <span class="kw"><a href="https://rdrr.io/r/graphics/abline.html">abline</a></span>(<span class="dt">h=</span>cutoffs[i], <span class="dt">lty=</span><span class="dv">2</span>)
}</code></pre></div>
<p><img src="c-metacell_annotation_files/figure-html/unnamed-chunk-17-1.png" width="480"></p>
<p>We define the annotations with their thresholds on genes in a tab-delimited file:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">marks_colors =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"guo2018_mc_partial_colorize.txt"</span>, <span class="dt">package =</span> <span class="st">"metacell"</span>), <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">h=</span>T, <span class="dt">stringsAsFactors=</span>F)
knitr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(marks_colors)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">group</th>
<th align="left">gene</th>
<th align="left">color</th>
<th align="right">priority</th>
<th align="right">T_fold</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">Treg</td>
<td align="left">FOXP3</td>
<td align="left">magenta</td>
<td align="right">1</td>
<td align="right">2.25</td>
</tr>
<tr class="even">
<td align="left">Tfh</td>
<td align="left">CXCL13</td>
<td align="left">darkolivegreen</td>
<td align="right">1</td>
<td align="right">16.00</td>
</tr>
<tr class="odd">
<td align="left">Naive</td>
<td align="left">TCF7</td>
<td align="left">navajowhite2</td>
<td align="right">1</td>
<td align="right">2.00</td>
</tr>
<tr class="even">
<td align="left">Prolif</td>
<td align="left">TOP2A</td>
<td align="left">lightblue</td>
<td align="right">1</td>
<td align="right">2.00</td>
</tr>
</tbody>
</table>
<p>The annotation name and its color are filled in the <em>group</em> and <em>color</em> columns. Threshold on each <em>gene</em> are in the <em>T_fold</em> column. Note that these are applied on the <span class="citation">@mc_fp</span> table, so should be fold change and not log2 fold-change. Each annotation can have several genes to define it, each with its own threshold. A metacell can match several rules, so the assigment might be ambigious. The <em>priority</em> value aims to address this. In the default mode, the lfp value of all genes passing their <em>T_fold</em> thresholds is multiplied by the gene priority and the gene with the maximal outcome is selected. If working in sequential coloring mode (setting configuration parameter <em>mcp_colorize_by_seq_priority</em> to be TRUE), genes are grouped by priority, and metacells are assigned to groups iteratively, starting with the genes in the first (smallest) priority, and only assigning groups to unassigned metacells in each iteration.</p>
<p>We use <em>mc_colorize</em> to annotate the metacelks with our gene rules, and plot the 2D projection to appreciate the result. Note that the <span class="citation">@color_key</span> table contains annotations, the plot has a legend. You can control its position with the <em>legend_pos</em> parameter (and control whether to show the legend at all with the logical configuration parameter <em>mcell_mc2d_plot_key</em>):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mc_colorize.html">mc_colorize</a></span>(<span class="dt">new_mc_id =</span> mc_id, <span class="dt">mc_id =</span> mc_non_ann_id, <span class="dt">marker_colors=</span>marks_colors)

<span class="kw"><a href="../reference/mcell_mc2d_plot.html">mcell_mc2d_plot</a></span>(<span class="dt">mc2d_id =</span> mc_id)
<span class="co">#&gt; png </span>
<span class="co">#&gt;   2</span></code></pre></div>
<pre><code>#&gt; [1] TRUE</code></pre>
<div class="figure">
<img src="lung_figs/guo2018_tpm_scaled_filt_Tumor_Normal_Blood_outClean.2d_graph_proj_col1.png" alt="Single-cell and metacells 2D projection" width="100%"><p class="caption">
Single-cell and metacells 2D projection
</p>
</div>
</div>
<div id="systematic-annotation-of-metacells" class="section level2">
<h2 class="hasAnchor">
<a href="#systematic-annotation-of-metacells" class="anchor"></a>Systematic annotation of metacells</h2>
<p>A more comprehensive and systematic approach to annotate metacells uses the metacells ‘confusion matrix’, a metacell pairwise similarity matrix, summarizing the K-nn graph connectivity between all cells in each pair of metacells. We start by hierarchically clustering the metacells (columns) of the confusion matrix:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mc_hc =<span class="st"> </span><span class="kw"><a href="../reference/mcell_mc_hclust_confu.html">mcell_mc_hclust_confu</a></span>(<span class="dt">mc_id =</span> mc_id, <span class="dt">graph_id =</span> graph_id)</code></pre></div>
<p>Next, we generate clusters of metacells (super-metacells, or sup_mc for convenience) based on this hierarchy with <em>mcell_mc_hierarchy</em>. These clusters are nodes in the hierarchical tree of metacells that their branch length is longer then <em>T_gap</em>. We visualize the confusion matrix and these clusters with <em>mcell_mc_plot_hierarchy</em>. Only sup_mcs with more than <em>min_nmc</em> metacells in them are plotted. The confusion matrix is shown at the bottom, and the top panel encodes the cluster hierarchy by showing the subtree of each sup_mc in blue, and the sibling subtree in gray.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">mc_sup =<span class="st"> </span><span class="kw"><a href="../reference/mcell_mc_hierarchy.html">mcell_mc_hierarchy</a></span>(<span class="dt">mc_id =</span> mc_id, <span class="dt">mc_hc =</span> mc_hc, <span class="dt">T_gap =</span> <span class="fl">0.04</span>)

<span class="kw"><a href="../reference/mcell_mc_plot_hierarchy.html">mcell_mc_plot_hierarchy</a></span>(<span class="dt">mc_id =</span> mc_id, <span class="dt">graph_id =</span> graph_id, 
                        <span class="dt">mc_order =</span> mc_hc<span class="op">$</span>order, 
                        <span class="dt">sup_mc =</span> mc_sup, 
                        <span class="dt">width =</span> <span class="dv">1200</span>, <span class="dt">height =</span> <span class="dv">2400</span>, 
                        <span class="dt">min_nmc=</span><span class="dv">2</span>, <span class="dt">show_mc_ids =</span> T)
<span class="co">#&gt; png </span>
<span class="co">#&gt;   2</span></code></pre></div>
<pre><code>#&gt; [1] TRUE</code></pre>
<div class="figure">
<img src="lung_figs/guo2018_tpm_scaled_filt_Tumor_Normal_Blood_outClean.supmc_confu_no_col.png" alt="Clustered metacells confusion matrix" width="100%"><p class="caption">
Clustered metacells confusion matrix
</p>
</div>
<p>There are several lists of enriched genes for each subtree. Let’s examine all the information supplied on a subree. For instance, subtree number 4 (which is clearly comprised of Treg metacells):</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(mc_sup[[<span class="dv">4</span>]])
<span class="co">#&gt; $marks</span>
<span class="co">#&gt;    GAPDH CD27-AS1    CXCR6   CARD16     SELL    CTLA4 TNFRSF1B     ICOS </span>
<span class="co">#&gt; 1.409033 1.425825 1.478982 1.527237 1.571408 1.609004 1.647381 1.729588 </span>
<span class="co">#&gt;  TNFRSF4    FOXP3     RGS1    TIGIT     CCR8     PIM2     SAT1     CD27 </span>
<span class="co">#&gt; 1.908381 1.955422 2.052494 2.065926 2.076183 2.108668 2.131010 2.170227 </span>
<span class="co">#&gt; TNFRSF18  MIR4632    IL2RA     BATF </span>
<span class="co">#&gt; 2.185034 2.188458 2.318691 2.542111 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $min_marks</span>
<span class="co">#&gt;     CXCR6      SELL      CD82      RGS1     SIRPG      ICOS      IL32 </span>
<span class="co">#&gt; 0.7418518 0.7487228 0.7501615 0.9337986 1.0267765 1.0475643 1.0644945 </span>
<span class="co">#&gt;    CARD16     CTLA4      CD74      CCR8     GAPDH  CD27-AS1     FOXP3 </span>
<span class="co">#&gt; 1.0738282 1.1351219 1.1766789 1.1937338 1.2025314 1.3105893 1.3563242 </span>
<span class="co">#&gt;     TIGIT      SAT1     IL2RA      PIM2      CD27      BATF </span>
<span class="co">#&gt; 1.5938838 1.6433828 1.7104313 1.7890192 2.0234116 2.1977205 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $marks_gap</span>
<span class="co">#&gt;      PKM     IFI6    IL1R2     ENO1     LDHA     SRGN     SAT1 TNFRSF1B </span>
<span class="co">#&gt; 1.173707 1.179759 1.208043 1.212875 1.219272 1.235725 1.239677 1.257179 </span>
<span class="co">#&gt; APOBEC3C    GAPDH     ICOS     CTSC  MIR4632    DDIT4  TNFRSF4     CCR8 </span>
<span class="co">#&gt; 1.298520 1.340427 1.342362 1.389729 1.482165 1.484738 1.512940 1.667040 </span>
<span class="co">#&gt;    CXCR6     RGS1 TNFRSF18     BATF </span>
<span class="co">#&gt; 1.860873 1.901821 1.971574 2.075978 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $marks_gap_anti</span>
<span class="co">#&gt;        GIMAP7        GIMAP4         TXNIP          CD52       S100A10 </span>
<span class="co">#&gt;    -2.1707354    -1.6104371    -1.5863388    -1.4286298    -1.1785400 </span>
<span class="co">#&gt;          FCMR GIMAP1-GIMAP5          SELL         LIME1       ALOX5AP </span>
<span class="co">#&gt;    -1.1444732    -1.1084550    -1.0938374    -1.0360991    -0.9999712 </span>
<span class="co">#&gt;           AES          LDHB         GSTK1        RPL13A          CCR7 </span>
<span class="co">#&gt;    -0.9894325    -0.9848750    -0.9780960    -0.9670300    -0.9525264 </span>
<span class="co">#&gt;        GIMAP5          RPS6        UBXN11       RARRES3          LEF1 </span>
<span class="co">#&gt;    -0.9523882    -0.9315339    -0.8861049    -0.8690605    -0.8554596 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $mcs</span>
<span class="co">#&gt; [1] 55 53 52 51 54 56</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $x_ord</span>
<span class="co">#&gt; [1] 6.5</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $sup_mcs</span>
<span class="co">#&gt;  [1] 40 38 39 58 55 53 52 51 54 56</span></code></pre></div>
<ul>
<li>
<em>mcs</em> - the ids of metacells in the subtree (those marked in blue)</li>
<li>
<em>sup_mcs</em> - the ids of metacells in the subtree and in the sibling subtree (those marked in blue and gray)</li>
<li>
<em>x_ord</em> - mean rank of the subtree metacells in the confusion matrix</li>
<li>
<em>marks</em> - averaging lfp over <em>mcs</em>, showing the top 20 genes</li>
<li>
<em>min_marks</em> - taking the minimal lfp value per gene over <em>mcs</em>, showing the top 20 genes</li>
<li>
<em>marks_gap</em> - showing the genes with maximal difference between the average lfp over <em>mcs</em> and average lfp over metacells in the sibling subtree <em>marks_gap_anti</em> - similar to <em>marks_gap</em> but showing genes with minimal difference (e.g. enriched in the sibling subtree)</li>
</ul>
<p>Top 5 <em>marks</em> genes are shown in the plot to the left of the sup_mc row, top 5 <em>marks_gap</em> are shown to the right, followed (after a ‘pipe’ sign) by the top 5 <em>marks_gap_anti</em> genes.</p>
<p>The annotation task based on this sup_mcs becomes the task of tagging sup-mcs, by browsing the figure and the more detailed sup_mc object, such that all metacells are covered by a tagged sup_mc. Annotation can also flow from bottom up, for instance, if we’re insterseted in annotating CD8+ ZNF683+ enriched metacells: let’s first see the lfp enrichment of both genes:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plt</span>(<span class="dt">gene1 =</span> <span class="st">'CD8A'</span>, <span class="dt">gene2 =</span> <span class="st">'ZNF683'</span>, <span class="dt">lfp =</span> lfp, <span class="dt">colors =</span> mc<span class="op">@</span>colors)</code></pre></div>
<p><img src="c-metacell_annotation_files/figure-html/unnamed-chunk-25-1.png" width="480"></p>
<p>We see that there are 3 metacells with ZNF683 lfp values above 2.5, but where are these metacells within the sup_mc tree? We define here a simple helper function named <em>query_sup_by_mcs</em>, that gets a list of metacells and reports the sup_mcs that contain them (the id of the sup_mc, number of queried metacells, number of queried metacells inside the sup_mc and the sup_mc size). We usually aim to find the minimal sup_mc that contains most of the metacells we’re interested in, in this case, sup_mc 42 seems like a reasonable choice:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">query_sup_by_mcs =<span class="st"> </span><span class="cf">function</span>(sup, mcs) 
{
    <span class="kw"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="st">'rbind'</span>, <span class="kw"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="dv">1</span><span class="op">:</span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(sup), <span class="cf">function</span>(i) { csup =<span class="st"> </span>sup[[i]]; <span class="kw"><a href="https://rdrr.io/r/base/data.frame.html">data.frame</a></span>(<span class="dt">id=</span>i, <span class="dt">n_mcs=</span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(mcs), <span class="dt">n_in=</span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(<span class="kw"><a href="https://dplyr.tidyverse.org/reference/setops.html">intersect</a></span>(mcs, csup<span class="op">$</span>mcs)), <span class="dt">sup_size=</span><span class="kw"><a href="https://rdrr.io/r/base/length.html">length</a></span>(csup<span class="op">$</span>mcs)) })) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/filter.html">filter</a></span>(n_in <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span>) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw"><a href="https://dplyr.tidyverse.org/reference/arrange.html">arrange</a></span>(sup_size)
}
<span class="kw">query_sup_by_mcs</span>(mc_sup, <span class="kw"><a href="https://rdrr.io/r/base/which.html">which</a></span>(lfp[<span class="st">'ZNF683'</span>, ] <span class="op">&gt;</span><span class="st"> </span><span class="fl">2.5</span>))
<span class="co">#&gt;   id n_mcs n_in sup_size</span>
<span class="co">#&gt; 1 43     3    2        4</span>
<span class="co">#&gt; 2 42     3    3        5</span>
<span class="co">#&gt; 3 40     3    3        7</span>
<span class="co">#&gt; 4 46     3    3       19</span>
<span class="co">#&gt; 5 44     3    3       28</span>
<span class="co">#&gt; 6 41     3    3       30</span>

<span class="kw"><a href="https://rdrr.io/r/base/print.html">print</a></span>(mc_sup[[<span class="dv">42</span>]])
<span class="co">#&gt; $marks</span>
<span class="co">#&gt;        CAPG       COTL1    APOBEC3C        PRF1       ITGAE     ALOX5AP </span>
<span class="co">#&gt;    1.741976    1.760421    1.774480    1.929533    2.148585    2.170378 </span>
<span class="co">#&gt;         ID2      ZNF683        CD8B       CXCR6        CTSW        GZMA </span>
<span class="co">#&gt;    2.242233    2.313298    2.344979    2.444342    2.776128    2.842893 </span>
<span class="co">#&gt;        CCL5 KLRC4-KLRK1        CCL4        GZMH       KLRK1        CD8A </span>
<span class="co">#&gt;    3.392455    3.481389    3.586614    3.655135    3.779911    3.999646 </span>
<span class="co">#&gt;        NKG7        GZMB </span>
<span class="co">#&gt;    4.393001    4.595379 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $min_marks</span>
<span class="co">#&gt;       ITGAE        CTSD    APOBEC3C    APOBEC3G      GPR171         ID2 </span>
<span class="co">#&gt;    1.012721    1.021769    1.029850    1.127767    1.148459    1.183031 </span>
<span class="co">#&gt;       COTL1        PRF1        CCL4        GZMA        CTSW       CXCR6 </span>
<span class="co">#&gt;    1.206467    1.368101    1.731148    1.834879    1.891544    2.115394 </span>
<span class="co">#&gt;        CD8B        CCL5 KLRC4-KLRK1        GZMH        GZMB       KLRK1 </span>
<span class="co">#&gt;    2.122480    2.879629    3.070701    3.145283    3.369617    3.434703 </span>
<span class="co">#&gt;        CD8A        NKG7 </span>
<span class="co">#&gt;    3.514828    3.649621 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $marks_gap</span>
<span class="co">#&gt;     UCP2    SIRPG    ITM2A HLA-DRB1     IDH2   CXCL13    DDIT4      LCK </span>
<span class="co">#&gt; 1.158478 1.168008 1.174318 1.197966 1.199905 1.203829 1.231473 1.268832 </span>
<span class="co">#&gt;   ARPC1B     GZMB     CD74   CORO1A APOBEC3C    COTL1 HLA-DRB6    ITGAE </span>
<span class="co">#&gt; 1.271518 1.277664 1.306401 1.334704 1.435025 1.450852 1.586647 1.645951 </span>
<span class="co">#&gt;    CXCR6     GZMA HLA-DRB5     GZMH </span>
<span class="co">#&gt; 1.681105 1.740864 1.799916 1.830240 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $marks_gap_anti</span>
<span class="co">#&gt;       IL7R      RPLP2    S100A10     LGALS1     GPR183     S100A4 </span>
<span class="co">#&gt; -1.6705130 -1.4081441 -1.3791475 -1.2715754 -1.1557306 -1.0069529 </span>
<span class="co">#&gt;      RPS12      ANXA1      RPS14    S100A11      RPL35       GNLY </span>
<span class="co">#&gt; -1.0064458 -0.9702105 -0.9231261 -0.9154989 -0.9065101 -0.9012424 </span>
<span class="co">#&gt;      RPL36     TMSB10      RPL30      RPS16      RPS29      RPS10 </span>
<span class="co">#&gt; -0.8934304 -0.8931744 -0.8871693 -0.8416471 -0.8384013 -0.8373846 </span>
<span class="co">#&gt;      RPL27      RPL32 </span>
<span class="co">#&gt; -0.8348400 -0.8348025 </span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $mcs</span>
<span class="co">#&gt; [1] 16 13 14 12 15</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $x_ord</span>
<span class="co">#&gt; [1] 54</span>
<span class="co">#&gt; </span>
<span class="co">#&gt; $sup_mcs</span>
<span class="co">#&gt; [1] 16 13 14 12 15  5  4</span></code></pre></div>
<p>The list of annotated sup_mcs should be in a tab-delimited table containing the id of the sup_mc, its name and color, as shown below. Note that the order of entries matters - the annotation function process this table sequentially, so the newest entry overrides previous ones (if a metacell below to several sup_mcs in the table). It is actually useful in some cases, where you first define a large group of metacells and then pick from within it a smaller group and re-annotate it.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">supmc_tab =<span class="st"> </span><span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"guo2018_supmc.txt"</span>, <span class="dt">package =</span> <span class="st">"metacell"</span>), <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">h=</span>T, <span class="dt">stringsAsFactors=</span>F)
knitr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(supmc_tab)</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="right">supid</th>
<th align="left">color</th>
<th align="left">name</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="right">2</td>
<td align="left">#107538</td>
<td align="left">Tfh</td>
</tr>
<tr class="even">
<td align="right">4</td>
<td align="left">#59b56f</td>
<td align="left">tumor-Treg</td>
</tr>
<tr class="odd">
<td align="right">8</td>
<td align="left">#a3ffb9</td>
<td align="left">blood-Treg</td>
</tr>
<tr class="even">
<td align="right">17</td>
<td align="left">#b3d4f0</td>
<td align="left">naive-TCF7</td>
</tr>
<tr class="odd">
<td align="right">27</td>
<td align="left">#6d8eaa</td>
<td align="left">naive-CXCR6</td>
</tr>
<tr class="even">
<td align="right">34</td>
<td align="left">#cee6b9</td>
<td align="left">CD4-cyto</td>
</tr>
<tr class="odd">
<td align="right">37</td>
<td align="left">#fdb913</td>
<td align="left">CD8-cyto</td>
</tr>
<tr class="even">
<td align="right">40</td>
<td align="left">#a76eaf</td>
<td align="left">CD8-dysf</td>
</tr>
<tr class="odd">
<td align="right">51</td>
<td align="left">#d29bc6</td>
<td align="left">CD8-GZMK</td>
</tr>
</tbody>
</table>
<p>Sometimes the sup_mcs will not cover all the metacells. Usually this is the case if we have rare cell types represented by a single metacell. For instance Mucosal associated invariant T cell (MAIT) cells in our data here:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">plt</span>(<span class="dt">gene1 =</span> <span class="st">'SLC4A10'</span>, <span class="dt">gene2 =</span> <span class="st">'KLRB1'</span>, <span class="dt">lfp =</span> lfp, <span class="dt">colors =</span> mc<span class="op">@</span>colors)</code></pre></div>
<p><img src="c-metacell_annotation_files/figure-html/unnamed-chunk-28-1.png" width="480"> We can annotate these metacells by defining thresholds on genes. Note that the <em>T_fold</em> threshold in this case is compared to the log2 lfp value. Processing of this table is also sequetial, so if a metacell matches several rules, the last one overwrites any of the previous rules, and coloring by this table will also override any coloring done based on the sup_mcs table.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">knitr<span class="op">::</span><span class="kw"><a href="https://rdrr.io/pkg/knitr/man/kable.html">kable</a></span>(<span class="kw"><a href="https://rdrr.io/r/utils/read.table.html">read.table</a></span>(<span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"guo2018_marks.txt"</span>, <span class="dt">package =</span> <span class="st">"metacell"</span>), <span class="dt">sep=</span><span class="st">"</span><span class="ch">\t</span><span class="st">"</span>, <span class="dt">h=</span>T, <span class="dt">stringsAsFactors=</span>F))</code></pre></div>
<table class="table">
<thead><tr class="header">
<th align="left">name</th>
<th align="left">gene</th>
<th align="left">color</th>
<th align="right">T_fold</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="left">CD8-dysf</td>
<td align="left">TUBA1B</td>
<td align="left">#a76eaf</td>
<td align="right">2</td>
</tr>
<tr class="even">
<td align="left">CD8-MAIT</td>
<td align="left">SLC4A10</td>
<td align="left">#ead37a</td>
<td align="right">1</td>
</tr>
<tr class="odd">
<td align="left">CD8-dysf-KLRC4</td>
<td align="left">KLRC4</td>
<td align="left">#6b3273</td>
<td align="right">3</td>
</tr>
<tr class="even">
<td align="left">naive-TCF7</td>
<td align="left">XIST</td>
<td align="left">#b3d4f0</td>
<td align="right">1</td>
</tr>
</tbody>
</table>
<p>Once both these tables are defined, we can colorize (e.g. annotate) metacells with them using the <em>mc_colorize_sup_hierarchy</em> function:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw"><a href="../reference/mc_colorize_sup_hierarchy.html">mc_colorize_sup_hierarchy</a></span>(<span class="dt">mc_id =</span> mc_id,
                          <span class="dt">supmc =</span> mc_sup,
                          <span class="dt">supmc_key =</span> <span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"guo2018_supmc.txt"</span>, <span class="dt">package =</span> <span class="st">"metacell"</span>),
                          <span class="dt">gene_key=</span> <span class="kw"><a href="https://rdrr.io/r/base/system.file.html">system.file</a></span>(<span class="st">"extdata"</span>, <span class="st">"guo2018_marks.txt"</span>, <span class="dt">package =</span> <span class="st">"metacell"</span>))


<span class="co"># generate metacell clusters heatmap</span>
<span class="kw"><a href="../reference/mcell_mc_plot_hierarchy.html">mcell_mc_plot_hierarchy</a></span>(<span class="dt">mc_id =</span> mc_id,
                                                <span class="dt">graph_id =</span> graph_id,
                                                <span class="dt">mc_order =</span> mc_hc<span class="op">$</span>order,
                                                <span class="dt">sup_mc =</span> mc_sup,
                                                <span class="dt">width=</span><span class="dv">1200</span>, <span class="dt">height=</span><span class="dv">2400</span>, <span class="dt">min_nmc=</span><span class="dv">2</span>, <span class="dt">show_mc_ids=</span> T)
<span class="co">#&gt; png </span>
<span class="co">#&gt;   2</span></code></pre></div>
<pre><code>#&gt; [1] TRUE</code></pre>
<div class="figure">
<img src="lung_figs/guo2018_tpm_scaled_filt_Tumor_Normal_Blood_outClean.supmc_confu_col.png" alt="Clustered metacells confusion matrix" width="100%"><p class="caption">
Clustered metacells confusion matrix
</p>
</div>
<p>And the updated 2D projection would look like this:</p>
<pre><code>#&gt; png 
#&gt;   2</code></pre>
<pre><code>#&gt; [1] TRUE</code></pre>
<div class="figure">
<img src="lung_figs/guo2018_tpm_scaled_filt_Tumor_Normal_Blood_outClean.2d_graph_proj_final_col.png" alt="Single-cell and metacells 2D projection" width="100%"><p class="caption">
Single-cell and metacells 2D projection
</p>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">

        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#introduction">Introduction</a></li>
      <li><a href="#setting-up">Setting up</a></li>
      <li><a href="#initial-exploration-of-the-metacells">Initial exploration of the metacells</a></li>
      <li><a href="#metacells-gene-enrichment-table">Metacells gene enrichment table</a></li>
      <li><a href="#manual-annotation-by-thresholding-gene-enrichment">Manual annotation by thresholding gene enrichment</a></li>
      <li><a href="#systematic-annotation-of-metacells">Systematic annotation of metacells</a></li>
      </ul>
</div>
      </div>

</div>



      <footer><div class="copyright">
  <p>Developed by <a href="http://compgenomics.weizmann.ac.il/tanay/">Amos Tanay</a>.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.4.0.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
